---
title: "MetagenomicAssemblyFettsirte"
output:
  html_document: default
  html_notebook: default
bibliography: W채denswil.bib
---
Understanding the full genomic potential within a microbial community is key to comprehend the biological/functional abilities of this specific meta community. Virulence, Resistence aswell as specific functions/phenotypes are likely dependend on specific aquired features which might not be present in reference genomes.

Knowning the composition of microbial communities is the basis for understanding interactions therein. Interactions between organisms can quickly become very complex. Therefore understanding the general patterns in lower complexity assemblages should enable a crucial insight into all different microbial interactions. As most functional asseys rely on a well sequenced and annotated genome, genome assembly and annotation is key for downstream analysis. Sofar most analysis aiming at understand micbrobial compostion have used a targeted amplicon approach. 

However there are well known limitations including biases introduced by copy-number variations [@Klappenbach2001], variability in amplification efficiency [@Engelbrektson2010], inconsistencies when targeting different regions of this gene [@Claesson2010], and problems with accurately and consistently delineating prokaryotic species[@Gevers2005].

Therefor to truely understand the specific species and intra-species diversity shotgun whole genome sequencing of the metagenome should be considered. The most recent improvments in third generation sequencing enable sofar unprecedented oppurtunities to sequence entire metagenomes of low complexity communities. 

Additonally to the taxonomic profiling whole metagenome sequencing has further advantages in the gene functional analysis [@Qin2012,@Forslund2013] aswell as the genomic variation analysis [@Sunagawa2013]

##advantages and difficulties of long read sequencing on solving metagenomes

Ongoing improvements of both Pacbio (RSII and Sequel) as well as Nanopore (Miniion and Promethion) for increased read length, higher throughput and read quality have changed the field of whole genome sequencing. Assembling complete circularized genomes is becoming a standard. It is well known that for completely assembling a genome the sequencing reads need to span the longest intragenomic repeat region [@Olson2017]. For metagenomic assemblies the reads however not only have to span intragenomic repeats but also intergenomic repeats. An intergenomic repeat is a homological sequence between two or more co-occurring organism. These repeats are not generally of low complexity but can also be of low divergence and therefore are intrinsically indistinguishable. These repeats can be resolved by either spanning the repeat with long sequences to anchor within unique flanking regions or by increased sequencing quality whereby genomes can be discriminated solely based on subtle differences.

##Advantages of whole genome metagenomics in comparison to targeted amplicon sequencing of communities.
Sequencing the entire DNA potential within the environment of interest has multiple advantages.
1. No PCR biases due to different binding abilities of the primers to different sequences.
2. Better resolution to identifing higher order taxonomical differences. For exmaple in pinpointing strain levela or resistence. Like antibiotic restistance. 
3. ability to indentify core metagenomes if multiple treatments and multiple replicates. 
4. possibility to identifiy treatment specific metagenomes
5. functional annotation of sequencing data and implications of genotype/genes on phenotype
6.Fundamental basis for all further 'omic analysis and functional analysis of specific environment. 
7. Without complete genomes of the specific strains, studies are limited to the study of targeted genes or SNPs. Final ability to study large structural variation (relocations and horizontal gene transfer), which potentially have sofar unprecedented effects on the phenotype. 


##Our study setup

We have ran two Fettsirte on one Sequel run each. These Fettsirte originate from the Gruyere cheese production in Switzerland. One Fettsirte is called Cottens and has approxamatly 2 Mio. reads. the second Fettsirte is called LaPraz and has approxamatly 5 Mio. reads. 

####Set global options and bash shortcuts

```{r global_options}
knitr::opts_chunk$set(fig.width=12, fig.height=8,warning=FALSE, message=FALSE)
```

##creat file shortcuts

Before starting all file and programm locations have to be filled out or check. 
These shortcut locations will be used throughout the code and therefor have to only be adjusted once. 

However the fileShortucts.sh script location has to be changed with a search/replace throughout the scirpt if necessary!

```{bash file_shortcuts,eval=FALSE}

file=/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

##header
echo '#!/bin/bash' > $file
##locations
echo 'DATA_ROOT=/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109' >> $file
echo 'minimap2_root=/home/bioinf/extradrv/minimap2/minimap2' >> $file
echo 'miniasm_root=/home/bioinf/extradrv/miniasm' >> $file
echo 'SMRTLINK_ROOT=/home/bioinf/extradrv/opt/pacbio/smrtlink/userdata/jobs_root/000' >> $file
echo 'blast_root=/home/bioinf/extradrv/ncbi-blast-2.7.1+/bin' >> $file
echo 'blastdb=/home/bioinf/bioinf_archive/43_sovi_storage/blastArchive' >> $file
echo 'sprai_root=/home/bioinf/extradrv/sprai-0.9.9.23' >> $file
echo 'abruijn_root=/home/bioinf/extradrv/abruijn/ABruijn-master/bin' >> $file
echo 'circular_root=/home/bioinf/extradrv/identifyCircularContigs-master' >> $file
echo 'bbmap_root=/opt/Geneious/bundledPlugins/com.biomatters.plugins.bbtools.BBToolsPlugin/com/biomatters/plugins/bbtools/BBMap_37.25/bbmap/' >> $file
echo 'blob_root=/home/bioinf/extradrv/blobology' >> $file
echo 'QUAL_ROOT=/home/bioinf/apps/qualimap_v2.2.1/' >> $file
echo 'mpa_dir=/home/bioinf/extradrv/biobakery-metaphlan2-d8ab9ca4244c' >> $file
echo 'aditiBlastDB=/home/bioinf/bioinf_archive/30_vaad_storage/BLASTDB/nt_db' >> $file
echo 'mOTU_ROOT=/home/bioinf/extradrv' >> $file
echo 'metapha=/home/bioinf/extradrv/biobakery-metaphlan2-d8ab9ca4244c' >> $file
echo 'ngmlr_root=/home/bioinf/extradrv/ngmlr-0.2.6' >> $file
echo 'sniffles_root=/home/bioinf/extradrv/Sniffles-master/bin/sniffles-core-1.0.8' >> $file
echo 'flye_root=/home/bioinf/extradrv/Flye/bin/flye' >> $file
echo 'pacbio_root=/home/bioinf/extradrv/opt/pacbio/smrtlink/smrtcmds/bin' >> $file
echo 'mauve_root=/home/bioinf/extradrv/mauve_snapshot_2015-02-13/linux-x64/' >> $file



echo 'ASSEMBLY_FLYE_POLISHED_FASTA=/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/15_flye/assembly_cottens_min5000/SmrtLink_polished_reference/reference_cottens.fasta' >> $file

#echo 'timestamp() {date +"%T"}'>> $file


```
##File locations


```{bash file_locations ,eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

##rawData
rawData_cottens=$DATA_ROOT/01_SMRTlink_data/cottens/reads.fastq
rawData_lapraz=$DATA_ROOT/01_SMRTlink_data/lapraz/reads.fastq

##size selected rawData
min5000bp_cottens=$DATA_ROOT/01_SMRTlink_data/01_cottens_min5000bp/reads.fastq
min5000bp_lapraz=$DATA_ROOT/01_SMRTlink_data/01_lapraz_min5000bp/reads.fastq

```


##Sequencing stats
The sequencing stats were obtained from the standard ouptut file of SMRTlink and are as following:


```{r packages, warning=FALSE, echo=TRUE,eval=TRUE}

.libPaths("/home/bioinf/bioinf_data/43_sovi/R/")

suppressPackageStartupMessages(library("ggplot2"))
suppressPackageStartupMessages(library("rjson"))
#suppressPackageStartupMessages(library("knitr"))
#source("https://bioconductor.org/biocLite.R")
#suppressPackageStartupMessages(library("Repitools"))
suppressPackageStartupMessages(library("curatedMetagenomicData"))
suppressPackageStartupMessages(library("fastqcr"))
suppressPackageStartupMessages(library("readr"))
suppressPackageStartupMessages(library("dplyr"))
#suppressPackageStartupMessages(library("MetaComp"))
suppressPackageStartupMessages(library("ggplot2"))
suppressPackageStartupMessages(library("gridExtra"))



source("/home/bioinf/bioinf_data/43_sovi/R_functions/R_functions/g_legend.R")

# biocLite("curatedMetagenomicData")
# biocLite("fastqcr")
```

##make Quality control of the raw reads
####BamQc
```{r bamqc_raw ,eval=FALSE}
##------------
##Definitions
##------------
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh
##------------
##bamqc for raw reads
##------------
for name in cottens lapraz
   do
    mkdir -p $DATA_ROOT/01_SMRTlink_data/${name}/bamqc
    $QUAL_ROOT/qualimap bamqc -bam /home/bioinf/bioinf_archive/32_scmi_storage/archivePacBio/fettsirte1_02112017/scratch/pacbio_tmp/tmpWL3QGn/1_8_La_praz_3_D02.tar.gz/m54073_171026_034928.subreads.bam -outdir $DATA_ROOT/01_SMRTlink_data/${name}/bamqc --java-mem-size=2G
   done
   

```


make FASTQC
```{bash fastqc , warning=FALSE, echo=FALSE,eval=FALSE}
##------------
##Definitions
##------------
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh
threads=4

##------------
##fastqc for raw reads
##------------
for name in cottens lapraz
   do
    mkdir -p $DATA_ROOT/01_SMRTlink_data/${name}/fastqc/unzipped
    fastqc $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fastq -t ${threads} -o $DATA_ROOT/01_SMRTlink_data/${name}/fastqc
    sed -n '1~4s/^@/>/p;2~4p' $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fastq > $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fasta
    seq_length.py $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fasta > $DATA_ROOT/01_SMRTlink_data/${name}/seq_length.ttxt
   done
   
##------------
##fastqc for min_5000bp reads
##------------
threads=10
for name in cottens lapraz
   do
    #mkdir -p $DATA_ROOT/01_SMRTlink_data/01_${name}_min5000bp/fastqc/unzipped
    fastqc $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_min5000bp.fastq -t ${threads} -o $DATA_ROOT/01_SMRTlink_data/${name}/fastqc
    #sed -n '1~4s/^@/>/p;2~4p' $DATA_ROOT/01_SMRTlink_data/01_${name}_min5000bp/reads_${name}_min5000.fastq > $DATA_ROOT/01_SMRTlink_data/01_${name}_min5000bp/reads_${name}_min5000.fasta
    #seq_length.py $DATA_ROOT/01_SMRTlink_data/01_${name}_min5000bp/reads_${name}_min5000.fasta > $DATA_ROOT/01_SMRTlink_data/01_${name}_min5000bp/seq_length.ttxt
   done 
   
##------------
##fastqc for preassembled reads 
##------------
#cp $DATA_ROOT/01_SMRTlink_data/cottens/preassembled_reads/preads4falcon.fasta $DATA_ROOT/01_SMRTlink_data/cottens/preassembled_reads/preads4falcon_cottens.fasta
#cp $DATA_ROOT/01_SMRTlink_data/lapraz/preassembled_reads/preads4falcon.fasta $DATA_ROOT/01_SMRTlink_data/lapraz/preassembled_reads/preads4falcon_lapraz.fasta

for name in cottens lapraz
   do
    mkdir -p $DATA_ROOT/01_SMRTlink_data/${name}/preassembled_reads/fastqc
    fastqc $DATA_ROOT/01_SMRTlink_data/${name}/preassembled_reads/preads4falcon_${names}.fasta -t ${threads} -o $DATA_ROOT/01_SMRTlink_data/${name}/preassembled_reads/fastqc
    seq_length.py $DATA_ROOT/01_SMRTlink_data/${name}/preassembled_reads/preads4falcon_${names}.fasta > $DATA_ROOT/01_SMRTlink_data/01_${name}_min5000bp/seq_length.ttxt
   done  
   
##------------
##fastqc for cicular consensus reads 
##------------
for name in cottens lapraz
   do
    mkdir -p $DATA_ROOT/01_SMRTlink_data/${name}_ccs/fastqc
    fastqc $DATA_ROOT/01_SMRTlink_data/${name}_ccs/ccs_${name}.fastq -t ${threads} -o $DATA_ROOT/01_SMRTlink_data/${name}_ccs/fastqc
    sed -n '1~4s/^@/>/p;2~4p' $DATA_ROOT/01_SMRTlink_data/${name}_ccs/ccs_${name}.fastq > $DATA_ROOT/01_SMRTlink_data/${name}_ccs/ccs_${name}.fasta
    seq_length.py $DATA_ROOT/01_SMRTlink_data/${name}_ccs/ccs_${name}.fasta > $DATA_ROOT/01_SMRTlink_data/${name}_ccs/seq_length.ttxt
   done    
 
 
 print(c("contig_1 :",        2251206-2251141))
 print(c("contig_7 :",       39877-39878))
print(c("scaffold_3 :",      1903347-1903107))

contig_1|arrow  2251141                                                                                                                                                                                                                                                    
contig_7|arrow  39878                                                                                                                                                                                                                                                          
scaffold_3|arrow        1903107   
   
```


```{r stats, warning=FALSE,eval=FALSE}
##----------
##fastqc
##----------
fastqc_cottens <- qc_read("/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/01_SMRTlink_data/cottens/fastqc/reads_fastqc.zip")
fastqc_lapraz <- qc_read("/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/01_SMRTlink_data/lapraz/fastqc/reads_fastqc.zip")

seq_length_cottens <- read_delim("~/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/01_SMRTlink_data/cottens/seq_length.ttxt", "\t", escape_double = FALSE, col_names = c("read","length"), trim_ws = TRUE)

seq_length_lapraz <- read_delim("~/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/01_SMRTlink_data/lapraz/seq_length.ttxt", "\t", escape_double = FALSE, col_names = c("read","length"), trim_ws = TRUE)

##reads longer than
for(i in c("cottens","lapraz")){
nam <- paste("readlength_distribution_",i,sep="")
nam_2 <- paste("fastqc_",i,sep="")
nam_3 <- paste("seq_length_",i,sep="")

ggplot(get(nam_3),aes(x=length))+geom_histogram(binwidth = 500)+theme_light()+labs(title=nam_3)

sizethreshold<- 7000
print(c("There are:",sum(get(nam_3)[,2]>sizethreshold),"reads larger than",sizethreshold,"for",i))
}

##reads range
for(i in c("cottens","lapraz")){
nam <- paste("readlength_distribution_",i,sep="")
nam_2 <- paste("fastqc_",i,sep="")
nam_3 <- paste("seq_length_",i,sep="")

print(c("For ",i,"the min read length is:",min(get(nam_3)[,2]),"and max: ",max(get(nam_3)[,2])))

}

##make a stats output 


json_file <- "../04_HGAP4/20171114assembly_lapraz/stats/mapping_stats_hgap.json"
general_stats_1 <- fromJSON(file=json_file)

json_file <- "../04_HGAP4/20171114assembly_cottens//stats/mapping_stats_hgap.json"
general_stats_2 <- fromJSON(file=json_file)

##create table

total_seq_bases_1 <- round((general_stats_1$attributes[[4]]$value)/(general_stats_1$attributes[[2]]$value*100)*100)
total_seq_bases_2 <- round((general_stats_2$attributes[[4]]$value)/(general_stats_2$attributes[[2]]$value*100)*100)



stats_table <- data.frame("total Seq bases"=c(total_seq_bases_1,total_seq_bases_2),
                          "percent mapped bases"=c(general_stats_1$attributes[[1]]$value,general_stats_2$attributes[[1]]$value),
                          "number of subread"=c(general_stats_1$attributes[[3]]$value,general_stats_2$attributes[[3]]$value),
                          "number of subreads bases (realigned)"=c(general_stats_1$attributes[[3]]$value,general_stats_2$attributes[[3]]$value),
                          "Subread Length Mean"=c(general_stats_1$attributes[[5]]$value,general_stats_2$attributes[[5]]$value),
                          "Subread Length N50"=c(general_stats_1$attributes[[6]]$value,general_stats_2$attributes[[6]]$value),
                          "Subread Length Max"=c(general_stats_1$attributes[[8]]$value,general_stats_2$attributes[[8]]$value)
                          )
rownames(stats_table) <- c("lapraz","cottens")
knitr::kable(stats_table)
```

#Multifastqc of all raw data 
In order to have an overview over the raw data, the min 5000bp reads and the preassembled reads we create in the following a common multifastqc report. 
```{bash mutlifastqc_01 ,eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

mkdir -p $DATA_ROOT/01_SMRTlink_data/MultiQc/
multiqc  -f --title mulitQC_fasta --outdir $DATA_ROOT/01_SMRTlink_data/MultiQc/ \
        $DATA_ROOT/01_SMRTlink_data/cottens/fastqc \
        $DATA_ROOT/01_SMRTlink_data/lapraz/fastqc \
        $DATA_ROOT/01_SMRTlink_data/01_cottens_min5000bp/fastqc \
        $DATA_ROOT/01_SMRTlink_data/01_lapraz_min5000bp/fastqc \
        $DATA_ROOT/01_SMRTlink_data/cottens_ccs/fastqc \
        $DATA_ROOT/01_SMRTlink_data/lapraz_ccs/fastqc 
```
#1. Taxon profiling from metagenomic data
Knowning the organsims within a community is often the first and for many studies the most important information. For shotgun metagenome studies this is often done before assembling the sequencing data, as only the most abundant species can be de novo assembled [@Driscoll2017]. The key advantage of the metagenomic approach over a targeted amplicon approach is mainly the high resolution. A palette of different methods have enabled to profile strain level diversity from metagenomic data. However this approach is often limited to the thoroughness of the sequence database.

Apart from a phylogenetic survey of the microbial communities the shotgun metagenomic approach enables the de novo assembly of the dominant species which was sofar mainly restricted to cultureable species.  Thereby enabling the analysis of structural variation, genomic linkage of functions to specific genes and organisms and building the necassary basis for further omics analysis of this microbial community. In future with studies including multiple replicates the theoretical understanding of microbial communities could change to a core metagenome approach idenpednent of specific organsims but rather on the basis of occuring genes or functions.

Further getting an idea of the metagenome size and genetic divergence is potentially crucial for the parametrisization of the assemblers.

##Obtions for species and strain profiling
1. Metaphlan2: enables the profiling based on the clade-specific-marker genes(genes computed from the constantly updated prokaryotic Refseqdatabase [@Segata2013].
2. mOTU: based on universal, single-copy marker genes which provide prokaryotic species boundaries at higher resolution than 16SrRNA [@Sunagawa2013]. However does not work with Pacbio raw reads.


###Metaphlan2

Metaphlan2 is a tool based on the blastdb and specifically of uniq genes for each taxonomic classes therein. 
However this method was orignally developed for the usage with Illumina short read technology. 
In the following we aim to test how well the taxonomic classification/profiling works for raw long read sequencing data. 

For the taxonmic profiling we use the preassemble reads. These reads are both longer aswell as error corrected which should increase the power of the results.  

```{bash metaphlan2 , warning=F, eval=F}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

##----------------
##metaphlan Usage for raw reads
##----------------

for name in cottens lapraz
   do
    mkdir -p $DATA_ROOT/06_metaphlan/${name}_raw/{bowtie2ouput,metaphlan_profile}/
    rm $DATA_ROOT/06_metaphlan/${name}_raw/bowtie2ouput/metagenome_${name}_raw.bowtie2.bz2
    $mpa_dir/metaphlan2.py $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fastq --bowtie2out $DATA_ROOT/06_metaphlan/${name}_raw/bowtie2ouput/metagenome_${name}_raw.bowtie2.bz2 --nproc 12 --input_type fastq > $DATA_ROOT/06_metaphlan/${name}_raw/metaphlan_profile/profiled_metagenome_${name}_raw.txt
   done
   
##----------------
##metaphlan Usage for preassembled reads
##----------------

for name in cottens lapraz
   do
    mkdir -p $DATA_ROOT/06_metaphlan/${name}_preassembled/{bowtie2ouput,metaphlan_profile}/
    rm $DATA_ROOT/06_metaphlan/${name}_preassembled/bowtie2ouput/metagenome_${name}_preassembled.bowtie2.bz2
    $mpa_dir/metaphlan2.py $DATA_ROOT/01_SMRTlink_data/${name}/preassembled_reads/preads4falcon_${name}.fasta --bowtie2out $DATA_ROOT/06_metaphlan/${name}_preassembled/bowtie2ouput/metagenome_${name}_preassembled.bowtie2.bz2 --nproc 12 --input_type fasta > $DATA_ROOT/06_metaphlan/${name}_preassembled/metaphlan_profile/profiled_metagenome_${name}_preassembled.txt
   done
   
```
From the Metaphlan2 output it seems as if only the three most common cheese species are present in the lapraz metagenome.

```{r metaphlan2_analysis , warning=FALSE,eval=FALSE}

install.packages("Metacomb")

metaphlan_cottens <- load_metaphlan_assignment("/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/06_metaphlan/cottens_preassembled/metaphlan_profile/profiled_metagenome_cottens_preassembled.txt")

metaphlan_lapraz <- load_metaphlan_assignment("/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/06_metaphlan/lapraz_preassembled/metaphlan_profile/profiled_metagenome_lapraz_preassembled.txt")

```

##PanPhlAn
PanPhlAn is a very recent species/strain profiler for shotgun metagenomic data. 

In a first step, they have a good manual how to batch download a bunch of genomes from NCBI. 

```{bash batchDownloadNCBI ,eval=FALSE,echo=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

# 1) download the complete list of available bacteria genomes
mkdir -p $DATA_ROOT/17_PanPhlAn/{Lhel,Ldel,Sthe}
location=$pwd
cd $DATA_ROOT/17_PanPhlAn
wget -e use_proxy=yes -e ftp_proxy=proxy-bvcol.admin.ch:8080  ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/bacteria/assembly_summary.txt
cd $location

# 2) extract all directory-names from file assembly_summary.txt related to a species, for example all links to Bacteroides ovatus

grep -E 'Lactobacillus helveticus' assembly_summary.txt | cut -f 20 > $DATA_ROOT/17_PanPhlAn/Lhel/lhev.txt
grep -E 'Lactobacillus delbrueckii' assembly_summary.txt | cut -f 20 > $DATA_ROOT/17_PanPhlAn/Ldel/Ldel.txt
grep -E 'Streptococcus thermophilus ' assembly_summary.txt | cut -f 20 > $DATA_ROOT/17_PanPhlAn/Sthe/Sthe.txt


# 3) create download scripts


awk 'BEGIN{FS=OFS="/";filesuffix="genomic.fna.gz"}{ftpdir=$0;asm=$10;file=asm"_"filesuffix;print "wget -e use_proxy=yes -e ftp_proxy=proxy-bvcol.admin.ch:8080 "ftpdir,file}' $DATA_ROOT/17_PanPhlAn/Lhel/lhev.txt > $DATA_ROOT/17_PanPhlAn/Lhel/lhev_download_fna_files.sh
awk 'BEGIN{FS=OFS="/";filesuffix="genomic.gff.gz"}{ftpdir=$0;asm=$10;file=asm"_"filesuffix;print "wget -e use_proxy=yes -e ftp_proxy=proxy-bvcol.admin.ch:8080 "ftpdir,file}' $DATA_ROOT/17_PanPhlAn/Lhel/lhev.txt  > $DATA_ROOT/17_PanPhlAn/Lhel/lhev_download_gff_files.sh

awk 'BEGIN{FS=OFS="/";filesuffix="genomic.fna.gz"}{ftpdir=$0;asm=$10;file=asm"_"filesuffix;print "wget -e use_proxy=yes -e ftp_proxy=proxy-bvcol.admin.ch:8080 "ftpdir,file}' $DATA_ROOT/17_PanPhlAn/Ldel/Ldel.txt > $DATA_ROOT/17_PanPhlAn/Ldel/Ldel_download_fna_files.sh
awk 'BEGIN{FS=OFS="/";filesuffix="genomic.gff.gz"}{ftpdir=$0;asm=$10;file=asm"_"filesuffix;print "wget -e use_proxy=yes -e ftp_proxy=proxy-bvcol.admin.ch:8080 "ftpdir,file}' $DATA_ROOT/17_PanPhlAn/Ldel/Ldel.txt > $DATA_ROOT/17_PanPhlAn/Ldel/Ldel_download_gff_files.sh

awk 'BEGIN{FS=OFS="/";filesuffix="genomic.fna.gz"}{ftpdir=$0;asm=$10;file=asm"_"filesuffix;print "wget -e use_proxy=yes -e ftp_proxy=proxy-bvcol.admin.ch:8080 "ftpdir,file}' $DATA_ROOT/17_PanPhlAn/Sthe/Sthe.txt > $DATA_ROOT/17_PanPhlAn/Sthe/Sthe_download_fna_files.sh
awk 'BEGIN{FS=OFS="/";filesuffix="genomic.gff.gz"}{ftpdir=$0;asm=$10;file=asm"_"filesuffix;print "wget -e use_proxy=yes -e ftp_proxy=proxy-bvcol.admin.ch:8080 "ftpdir,file}' $DATA_ROOT/17_PanPhlAn/Sthe/Sthe.txt > $DATA_ROOT/17_PanPhlAn/Sthe/Sthe_download_gff_files.sh



# run download
source $DATA_ROOT/17_PanPhlAn/Lhel/lhev_download_fna_files.sh
source $DATA_ROOT/17_PanPhlAn/Lhel/lhev_download_gff_files.sh 

source $DATA_ROOT/17_PanPhlAn/Ldel/Ldel_download_fna_files.sh
source $DATA_ROOT/17_PanPhlAn/Ldel/Ldel_download_gff_files.sh

source $DATA_ROOT/17_PanPhlAn/Sthe/Sthe_download_fna_files.sh
source $DATA_ROOT/17_PanPhlAn/Sthe/Sthe_download_gff_files.sh 



ls
assembly_summary.txt
 download_dbff_files.sh
 download_fna_files.sh
 ftpdirpaths
 GCF_000178275.1_ASM17827v1_genomic.fna.gz
 GCF_000178275.1_ASM17827v1_genomic.gff.gz
 GCF_000218325.1_Bact_ovat_3_8_47FAA_V1_genomic.fna.gz
 GCF_000218325.1_Bact_ovat_3_8_47FAA_V1_genomic.gff.gz
 GCF_000273195.1_Bact_ovat_CL02T12C04_V1_genomic.fna.gz
 GCF_000273195.1_Bact_ovat_CL02T12C04_V1_genomic.gff.gz
 GCF_000273215.1_Bact_ovat_CL03T12C18_V1_genomic.fna.gz
 GCF_000273215.1_Bact_ovat_CL03T12C18_V1_genomic.gff.gz
 GCF_000699665.1_ASM69966v1_genomic.fna.gz
 GCF_000699665.1_ASM69966v1_genomic.gff.gz
 GCF_000699725.1_ASM69972v1_genomic.fna.gz
 GCF_000699725.1_ASM69972v1_genomic.gff.gz



# uncompress .gz files
gzip -d *.gz



# rename fna files: keep only GCF-number (remove all behind first dot)
for FILE in *.fna; do 
 mv ${FILE} ${FILE%%.*}.fna
done;



ls 
GCF_000178275.fna
GCF_000218325.fna
째째째
 


How to get the .ffn gene sequence files?

Update (September 2017)

PanPhlAn accepts .gff files as input (.ffn gene sequences are extracted by PanPhlAn)


panphlan_pangenome_generation.py -c speciesname --i_fna ncbi_download_fna/ --i_gff ncbi_download_gff/ -o database/



Prokka (recommended)
 You can run Prokka on the genome files (.fna) to re-annotate all genomes by using the same approach. The Prokka .gff files can be used as input for PanPhlAn
http://www.metagenomics.wiki/tools/annotation/prokka





panphlan_pangenome_generation.py -c speciesname --i_gff prokka_gff/ -o database/


```



##mOTU taxonomic annotation
In order to test the Metaphlan output we use a second well known taxonomic profiler for metagenomic datasets.
However mOTU does not work with Pacbio data. mOTU was made only for short next generation sequencing technologies, specifically for Illumina, Ion Torrent or 454. 

##blast raw reads against the entire ncbi database
Blasting all RawReads against the ncbi database should give the most comprehensive taxonomic analysis of the present data. 
This is however a very time consuming step. Especially as the we need to reduce the wordsize of the blast to 12 in order to account for the high error rate of the pacbio RawReads.

```{bash blastnrawRead , eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

##---------------- 
##fastq2fasta
##----------------
for name in cottens lapraz
   do
    SECONDS=0
    sed -n '1~4s/^@/>/p;2~4p' $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fastq  > $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fasta
    echo "RUNTIME (h:min:s)= $(($SECONDS / 3600)):$((($SECONDS / 60) % 60)):$(($SECONDS % 60))"
    echo END = `date +"%Y-%m-%d %H:%M:%S"`
   done
##---------------- 
##blast
##----------------

for name in cottens lapraz
   do
    mkdir -p $DATA_ROOT/01_SMRTlink_data/${name}/blast
    SECONDS=0
    $blast_root/blastn -num_threads 16 -max_hsps 1 -max_target_seqs 1 -task megablast -show_gis -query $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fasta -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle slen" -db $aditiBlastDB/nt -out $DATA_ROOT/01_SMRTlink_data/${name}/blast/${name}_blast_allRawReads.txt -evalue 0.01 -word_size 12
    echo "RUNTIME (h:min:s)= $(($SECONDS / 3600)):$((($SECONDS / 60) % 60)):$(($SECONDS % 60))"
    echo END $name = `date +"%Y-%m-%d %H:%M:%S"`
  done
  
##---------------- 
##blast batch run
##----------------
date=20170124

name=cottens
    #mkdir -p $DATA_ROOT/01_SMRTlink_data/${name}/blast
    #SECONDS=0
    
    LastLine=$(tail -1 $DATA_ROOT/01_SMRTlink_data/${name}/blast/${name}_blast_allRawReads.txt|cut -f 1)
    LineTrue=$(grep "$LastLine" -n $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fasta|awk -F":" '{print 758929-$1}')
    tail -$LineTrue $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fasta > $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw_subset_${date}.fasta
date=20170125

    $blast_root/blastn -num_threads 16 -max_hsps 1 -max_target_seqs 1 -task megablast -show_gis -query $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw_subset_${date}.fasta -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle slen" -db $aditiBlastDB/nt -out $DATA_ROOT/01_SMRTlink_data/${name}/blast/${name}_blast_allRawReads_subset_${date}.txt -evalue 0.01 -word_size 12
    echo "RUNTIME (h:min:s)= $(($SECONDS / 3600)):$((($SECONDS / 60) % 60)):$(($SECONDS % 60))"
    echo END $name = `date +"%Y-%m-%d %H:%M:%S"`
    

##---------------- 
##blast batch run reverse (michis machine)
##----------------

#tail -367062 $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fasta > $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw_master_michi_${date}.fasta

cp -r /home/bioinf/bioinf_archive/30_vaad_storage/BLASTDB/nt_db  /home/bioinf/bioinf_archive/30_vaad_storage/BLASTDB/nt_db_duplicate
blast_root_michi=/home/bioinf/Software/blast_2_7_1/ncbi-blast-2.7.1+/bin/blast/blastn
DATA_ROOT=/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109
aditiBlastDB=/home/bioinf/bioinf_archive/30_vaad_storage/BLASTDB/nt_db_dupliate


date=20170125
name=cottens
    #mkdir -p $DATA_ROOT/01_SMRTlink_data/${name}/blast
    #SECONDS=0
    
    #LastLine=$(tail -1 $DATA_ROOT/01_SMRTlink_data/${name}/blast/${name}_blast_allRawReads.txt|cut -f 1)
    #LineTrue=$(grep "$LastLine" -n $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fasta|awk -F":" '{print 758929-$1}')
    #tail -$LineTrue $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fasta > $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw_subset_${date}.fasta
    
    $blast_root_michi/blastn -num_threads 16 -max_hsps 1 -max_target_seqs 1 -task megablast -show_gis -query $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw_master_michi_${date}.fasta -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle slen" -db $aditiBlastDB/nt -out $DATA_ROOT/01_SMRTlink_data/${name}/blast/${name}_blast_allRawReads_reverse_subset_${date}.txt -evalue 0.01 -word_size 12
    #echo "RUNTIME (h:min:s)= $(($SECONDS / 3600)):$((($SECONDS / 60) % 60)):$(($SECONDS % 60))"
    #echo END $name = `date +"%Y-%m-%d %H:%M:%S"`
  

```

#Assemblies
The subreads are then used to assemble the genomes. In this point lies the major advantage of long reads. Long reads enable the spanning of repeat regions. In general long read assemblers consist of a Overlap Layout Consensus (OLC) algorithm. 

In the following we try different assemblers. 

##HGAP4
HGAP4 is the pacbio assembler implemented in SMRTlink 5.0. In the prelimary analysis based on the three Lactobacillus helveticus strains HGAP4 showed the best performance (precision and recall) for metagenome data. the HGAP4 assembly was run on default with an expected overall genome size of 10Mb.
This was done in the SMRTlink GUI.

```{bash hgap4_circularity, eval =F}

source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

###==========================
##check circularity
###==========================
for name in cottens lapraz
  do 
    mkdir -p $DATA_ROOT/04_HGAP4/20171114assembly_${name}/fasta/multifasta
    $sprai_root/check_circularity.pl $DATA_ROOT/04_HGAP4/20171114assembly_${name}/fasta/assembly.fasta $DATA_ROOT/04_HGAP4/20171114assembly_${name}/fasta/multifasta --force > $DATA_ROOT/04_HGAP4/20171114assembly_${name}/fasta/cirularity.txt
  done


```
##flye
Flye is the updated version of the Abruijn assembler.It is originally made to optimise repeat rich genome assemblies. However as with any non-pacbio based assembler (HGap) it can not handle the quality score and therefor polishing of the reads has to be done with quiver or arrow
The flye assembler creates a contigs.fasta and also an additional extended scaffolds.fasta file where it stitches together uncertain contigs.
We create for both wheey cultueres two assemblies. 
1. Starting the assembly with all raw subreads without quality nor length cutoff.
2. Starting the assembly with only raw subreads > 5000bp
```{bash flye_assembly , eval=FALSE}
##---------
##directories
##---------
##source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

##michis directories
flye_root=/home/bioinf/Software/Flye_2_31/Flye/bin/flye
blastn_root=blastn
minimap2_root=/home/bioinf/Software/minimap2Dec06/minimap2
DATA_ROOT=/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109
aditiBlastDB=/home/bioinf/bioinf_archive/30_vaad_storage/BLASTDB/nt_db/
##---------------
##creat flye directories
##---------------
for name in cottens lapraz
   do
    mkdir -p $DATA_ROOT/15_flye/{assembly_${name}_raw,assembly_${name}_min5000}/{quast,fasta,blast,rawReads2assembly_minimap2}
   done

##---------------
##create min5000bp reads
##---------------
for name in cottens lapraz
   do
    prinseq-lite.pl -fastq $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fastq -min_len 5000 -out_good $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_min5000bp -out_bad null -out_format 4 
   done



##---------------
##flye assembly with rawReads
##---------------
size=5m
for name in cottens lapraz
  do
    $flye_root -t 25 --pacbio-raw $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fastq --out-dir $DATA_ROOT/15_flye/assembly_${name}_raw/fasta --genome-size ${size} -m 5000 -i 3
    size=7m
  done
  
##---------------
##flye assembly with min5000bp reads
##---------------
size=5m
for name in cottens lapraz
  do
    $flye_root -t 25 --pacbio-raw $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_min5000bp.fastq --out-dir $DATA_ROOT/15_flye/assembly_${name}_min5000/fasta --genome-size ${size} -m 5000 -i 3
    size=7m
  done
  

```

####Minimap2 mapping of raw reads to the assemblies
We first need to mapped the raw reads or in general the reads which we want to use for polishing back to the reference genome proberly polish the assembled contigs. Pacbio offers a own aligner for this step. The PBAligner is describe [here](https://github.com/ben-lerch/Resequencing-3.0). This Pacbio aligner is not very well known. However it is necessary as it can incorporate the pulse width quality score into the alignment. ![PacBio polishing is performed in two stages: PBAlign, and variantCaller.](/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/99_Extras/Consensus_PB.png)

```{bash PBAlign_rawReadBam ,eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh
##-------   
##rawRead PBAlign to rawRead Scaffolds assembly
##-------

##cottens
RawBamSubreadscottens=/home/bioinf/bioinf_archive/32_scmi_storage/archivePacBio/fettsirte1_02112017/scratch/pacbio_tmp/tmpWL3QGn/2_7_Cottens_4_E02.tar.gz/m54073_171026_135911.subreads.bam
for name in cottens lapraz
   do 
      mkdir -p $DATA_ROOT/15_flye/assembly_${name}_raw/rawReads2assembly_PBAlign
  $pacbio_root/pbalign --nproc 12 --concordant --hitPolicy randombest --minAccuracy 70 --minLength 50 --algorithmOptions "--minMatch 12 --bestn 10 --minPctSimilarity 70.0 --refineConcordantAlignments" \
      --unaligned $DATA_ROOT/15_flye/assembly_${name}_raw/rawReads2assembly_PBAlign/raw_${name}_PBAaligned_Scaffolds_woConsensu_unaligned.bam \
      ${RawBamSubreadscottens} \
      $DATA_ROOT/15_flye/assembly_${name}_raw/fasta/scaffolds.fasta \
      $DATA_ROOT/15_flye/assembly_${name}_raw/rawReads2assembly_PBAlign/raw_${name}_PBAaligned_Scaffolds_woConsensu.bam 
  

      ##lapraz
      RawBamSubreadscottens=/home/bioinf/bioinf_archive/32_scmi_storage/archivePacBio/fettsirte1_02112017/scratch/pacbio_tmp/tmpWL3QGn/1_8_La_praz_3_D02.tar.gz/m54073_171026_034928.subreads.bam 
   done



##-------   
##rawRead PBAlign to min5000 Scaffolds assembly
##-------

##cottens
RawBamSubreadscottens=/home/bioinf/bioinf_archive/32_scmi_storage/archivePacBio/fettsirte1_02112017/scratch/pacbio_tmp/tmpWL3QGn/2_7_Cottens_4_E02.tar.gz/m54073_171026_135911.subreads.bam
for name in cottens lapraz
   do 
      mkdir -p $DATA_ROOT/15_flye/assembly_${name}_min5000/rawReads2assembly_PBAlign
  $pacbio_root/pbalign --nproc 12 --concordant --hitPolicy randombest --minAccuracy 70 --minLength 50 --algorithmOptions "--minMatch 12 --bestn 10 --minPctSimilarity 70.0 --refineConcordantAlignments" \
      --unaligned $DATA_ROOT/15_flye/assembly_${name}_min5000/rawReads2assembly_PBAlign/raw_${name}_PBAaligned_Scaffolds_woConsensu_unaligned.bam \
      ${RawBamSubreadscottens} \
      $DATA_ROOT/15_flye/assembly_${name}_min5000/fasta/scaffolds.fasta \
      $DATA_ROOT/15_flye/assembly_${name}_min5000/rawReads2assembly_PBAlign/raw_${name}_PBAaligned_Scaffolds_woConsensu.bam 
  

     

      ##lapraz
      RawBamSubreadscottens=/home/bioinf/bioinf_archive/32_scmi_storage/archivePacBio/fettsirte1_02112017/scratch/pacbio_tmp/tmpWL3QGn/1_8_La_praz_3_D02.tar.gz/m54073_171026_034928.subreads.bam 
      
      
   done

```


##Polishing with SmrtLink resequencing
After the Flye assembly the scaffolds need to be polished with the Pacbio resquencing pipeline implemented within SmrtLink. This has to be done as the fastq files do not actually contain per base quality values. The original unmapped bam file however has certain metadata which is internally used for quality assessement.


```{bash alginingForConsensusPolishing ,eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

##-------   
##rawRead Consensuspolishing of rawReads Scaffolds assembly (default)
##-------  


for name in cottens lapraz
   do 
    mkdir -p $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/assembly/
    
     samtools faidx $DATA_ROOT/15_flye/assembly_${name}_raw/fasta/scaffolds.fasta
    
    $pacbio_root/variantCaller --algorithm arrow --minConfidence 40 --minCoverage 5 --numWorkers 5 \
      -o $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/assembly/reference_${name}.fasta \
      -o $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/assembly/reference_${name}.fastq.gz \
      -o $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/assembly/reference_${name}_variants.vcf \
      -r $DATA_ROOT/15_flye/assembly_${name}_raw/fasta/scaffolds.fasta \
      $DATA_ROOT/15_flye/assembly_${name}_raw/rawReads2assembly_PBAlign/raw_${name}_PBAaligned_Scaffolds_woConsensu.bam 
      
    done
##-------   
##rawRead Consensuspolishing of min5000 Scaffolds assembly (default)
##-------  


for name in cottens lapraz
   do 
    mkdir -p $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/assembly/
    
    samtools faidx $DATA_ROOT/15_flye/assembly_${name}_min5000/fasta/scaffolds.fasta 
    
    $pacbio_root/variantCaller --algorithm arrow --minConfidence 40 --minCoverage 5 --numWorkers 5 \
      -o $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/assembly/reference_${name}.fasta \
      -o $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/assembly/reference_${name}.fastq.gz \
      -o $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/assembly/reference_${name}_variants.vcf \
      -r $DATA_ROOT/15_flye/assembly_${name}_min5000/fasta/scaffolds.fasta \
      $DATA_ROOT/15_flye/assembly_${name}_min5000/rawReads2assembly_PBAlign/raw_${name}_PBAaligned_Scaffolds_woConsensu.bam 

    done


  
```


####Blast assemblies
In the following we blasted both the contigs.fasta aswell as the scaffolds.fasta file against the prokaryotic ncbi database
```{bash blast_flye_assemblies ,eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

###-----------------
##blast rawread flye scaffold assembly (entire procaryotic ncbi database)
###-----------------

for name in cottens lapraz
   do
     mkdir -p $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/{blast,rawReads2assembly_minimap2,rawReads2assembly_ngml}
     blastn -num_threads 2 -max_hsps 1 -max_target_seqs 1 -task megablast -show_gis -query $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/assembly/reference_${name}.fasta -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle slen" -db $aditiBlastDB/nt -out $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/blast/blastAssemblyaginstScaffolds.txt -evalue 0.01 -word_size 64
   done

###-----------------
##blast min5000 flye assembly (entire procaryotic ncbi database)
###-----------------

for name in cottens lapraz
   do
     mkdir -p $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/{blast,rawReads2assembly_minimap2,rawReads2assembly_ngml}
     blastn -num_threads 15 -max_hsps 10 -max_target_seqs 10 -task megablast -show_gis -query $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/assembly/reference_${name}.fasta -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle slen" -db $aditiBlastDB/nt -out $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/blast/blastAssemblyaginstScaffolds.txt -evalue 0.01 -word_size 64
   done
```


####Minimap2 mapping of raw reads to the assemblies
```{bash minmap2_rawreads_flye , eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

   
##-------   
##rawRead mappingminimap2 to rawRead Scaffolds assembly
##-------   
for name in cottens lapraz
   do
     ${minimap2_root} -t 12 -ax map-pb $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/assembly/reference_${name}.fasta $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fastq |samtools sort -@8 -O BAM -o $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_Scaffolds_sorted.bam - 
   
   done 

##-------   
##rawRead mappingminimap2 to min5000 Scaffolds assembly
##-------   
for name in cottens lapraz
   do
     ${minimap2_root} -t 12 -ax map-pb $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/assembly/reference_${name}.fasta $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fastq |samtools sort -@8 -O BAM -o $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_Scaffolds_sorted.bam - 
   
   done   
   
```


####BamQC of flye assembly

```{bash bamqc_flye, eval=FALSE} 
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

##-------   
##bamqc mappingminimap2 to rawRead assembly (scaffolds)
##-------   
for name in cottens lapraz
   do
      mkdir -p $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/bamqc_scaffolds

      samtools view -b -f 4 $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_Scaffolds_sorted.bam > $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_sorted_Scaffolds_unampped.bam
      
     # bedtools bamtofastq -i $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/aligned_sorted_unmapped.bam -fq $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/aligned_sorted_unmapped.fastq
      
      $QUAL_ROOT/qualimap bamqc -bam $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_Scaffolds_sorted.bam -outdir $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/bamqc_scaffolds --java-mem-size=2G
      
      grep "Globals" -A 12  $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/bamqc_scaffolds/genome_results.txt |tail -11|sed '/^\s*$/d'|sed 's/=/\t/g'|sed -e 's/(.*%)//'|sed 's/,//g' > $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/bamqc_scaffolds/global_stats.txt
      
      grep "Coverage per contig" -A 1000  $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/bamqc/genome_results.txt > $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/bamqc/contig_stats.txt
 
      samtools index $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_Scaffolds_sorted.bam
  done  
  
##-------   
##bamqc mappingminimap2 to rawRead assembly (scaffolds)
##-------   
for name in cottens lapraz
   do
      mkdir -p $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/bamqc_scaffolds

      samtools view -b -f 4 $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_Scaffolds_sorted.bam > $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_sorted_Scaffolds_unampped.bam
      
     # bedtools bamtofastq -i $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/aligned_sorted_unmapped.bam -fq $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/aligned_sorted_unmapped.fastq
      
      $QUAL_ROOT/qualimap bamqc -bam $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_Scaffolds_sorted.bam -outdir $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/bamqc_scaffolds --java-mem-size=2G
      
      grep "Globals" -A 12  $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/bamqc_scaffolds/genome_results.txt |tail -11|sed '/^\s*$/d'|sed 's/=/\t/g'|sed -e 's/(.*%)//'|sed 's/,//g' > $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/bamqc_scaffolds/global_stats.txt
      
      grep "Coverage per contig" -A 1000  $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/bamqc/genome_results.txt > $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/bamqc/contig_stats.txt
 
      samtools index $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_Scaffolds_sorted.bam
  done   
```
####ngmlr/sniffles
Ngmlr and sniffles are mappers which are well known to identify structural variation
```{bash ngmlrSniffles, eval=FALSE}

source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

##-------   
##NGMLR of rawreads
##-------   
for name in cottens lapraz
   do
#    mkdir -p $DATA_ROOT/15_flye/assembly_${name}_raw/rawReads2assembly_ngmlr/
    $ngmlr_root/ngmlr -t 12 -r $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/assembly/reference_${name}.fasta -q $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fastq |samtools sort -@8 -O BAM -o $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/rawReads2assembly_ngmlr/raw_${name}_ngmlrMapped_Scaffolds_sorted.bam - 
  done
  
##-------   
##NGMLR of min5000
##-------   
for name in cottens lapraz
   do
#    mkdir -p $DATA_ROOT/15_flye/assembly_${name}_min5000/rawReads2assembly_ngmlr/
    $ngmlr_root/ngmlr -t 12 -r $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/assembly/reference_${name}.fasta -q $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fastq |samtools sort -@8 -O BAM -o $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/rawReads2assembly_ngmlr/raw_${name}_ngmlrMapped_Scaffolds_sorted.bam - 
  done
    


```

##Structural Variation
In a previous step we observed that most likely we will not have major repeats within the three expected species (Lhel, Ldel, Sthe). However we might have natural variation between the indiviual organism. This natural variation is most often in form of SingleNucleotideVariation (SNV). However we will also have structural variation possible hindering a finished assembly. Most likely, Structural variation outnumber SNV if viewed from the perspective of amount of basepairs effected as has been shown for humans [@Huddleston2017].   ![Structural Variation vs. Single nucleotide variation.](/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/99_Extras/SVvsSNV.png)

```{bash sniffels_flye, eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

##-------   
##sniffles of rawreads
##-------   
for name in cottens lapraz
   do
    mkdir -p $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/rawReads2assembly_ngmlr/sniffles
    $sniffles_root/sniffles -t 10 -m $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/rawReads2assembly_ngmlr/raw_${name}_ngmlrMapped_Scaffolds_sorted.bam -v $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/rawReads2assembly_ngmlr/sniffles/sniffles_${name}_raw.vcf

  done

##-------   
##sniffles of min5000
##-------   
for name in cottens lapraz
   do
    mkdir -p $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/rawReads2assembly_ngmlr/sniffles
    $sniffles_root/sniffles -t 10 -m $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/rawReads2assembly_ngmlr/raw_${name}_ngmlrMapped_Scaffolds_sorted.bam -v $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/rawReads2assembly_ngmlr/sniffles/sniffles_${name}_raw_min5000.vcf

  done

```

####Nucmer
In order to identify possible repeating regions between and within the genomes

```{bash nucmer_flye ,eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh
###==========================
##nucmer rawreads
###==========================

for name in cottens lapraz
   do
    mkdir -p $DATA_ROOT/15_flye/assembly_${name}_raw/nucmer/deltaFiltered
    
    cd $DATA_ROOT/15_flye/assembly_${name}_raw/nucmer/
    #nucmer -maxmatch -nosimplify $DATA_ROOT/15_flye/assembly_${name}_raw/fasta/scaffolds.fasta $DATA_ROOT/15_flye/assembly_${name}_raw/fasta/scaffolds.fasta 
    cd $DATA_ROOT

    #delta-filter -i 85 -l 000  $DATA_ROOT/15_flye/assembly_${name}_raw/nucmer/out.delta > $DATA_ROOT/15_flye/assembly_${name}_raw/nucmer/deltaFiltered/filtered.txt
    grep "[[:space:]]" $DATA_ROOT/15_flye/assembly_${name}_raw/nucmer/deltaFiltered/filtered.txt | grep -v "/" > $DATA_ROOT/15_flye/assembly_${name}_raw/nucmer/deltaFiltered/filtered_cleaned.txt
  done
  
###==========================
##nucmer min5000
###==========================

for name in cottens lapraz
   do
    mkdir -p $DATA_ROOT/15_flye/assembly_${name}_min5000/nucmer/deltaFiltered
    
    cd $DATA_ROOT/15_flye/assembly_${name}_min5000/nucmer/
    #nucmer -maxmatch -nosimplify $DATA_ROOT/15_flye/assembly_${name}_min5000/fasta/scaffolds.fasta $DATA_ROOT/15_flye/assembly_${name}_min5000/fasta/scaffolds.fasta 
    cd $DATA_ROOT

    #delta-filter -i 85 -l 3000  $DATA_ROOT/15_flye/assembly_${name}_min5000/nucmer/out.delta > $DATA_ROOT/15_flye/assembly_${name}_min5000/nucmer/deltaFiltered/filtered.txt
    grep "[[:space:]]" $DATA_ROOT/15_flye/assembly_${name}_min5000/nucmer/deltaFiltered/filtered.txt | grep -v "/" > $DATA_ROOT/15_flye/assembly_${name}_min5000/nucmer/deltaFiltered/filtered_cleaned.txt

  done
  
  
```

####assembling non mapping reads (plasmid assembly)
The reads which did not map with minimap2 to the flye assembled de novo reference genome are now taken and reassembled in a second stage. 
```{bash plasmidAssemblyFlye , eval =FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh
##---------
##plasmid assembly from rawReads
##---------

size=1m
for name in cottens lapraz
   do
      mkdir -p $DATA_ROOT/15_flye/assembly_${name}_raw/plasmidAssembly_flye/{fasta,blast,RawReads2plasmidassembly_minimap2}
      bedtools bamtofastq -i $DATA_ROOT/15_flye/assembly_${name}_raw/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_sorted_Scaffolds_unampped.bam -fq $DATA_ROOT/15_flye/assembly_${name}_raw/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_sorted_Scaffolds_unampped.fastq

      $flye_root -t 7 --pacbio-raw $DATA_ROOT/15_flye/assembly_${name}_raw/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_sorted_Scaffolds_unampped.fastq --out-dir $DATA_ROOT/15_flye/assembly_${name}_raw/plasmidAssembly_flye/fasta --genome-size ${size} -m 5000 -i 3 |tee $DATA_ROOT/15_flye/assembly_${name}_raw/plasmidAssembly_flye/fasta/assemblylog.txt
    size=1m
  done

##---------
##plasmid assembly from min5000
##---------
size=1m
for name in cottens lapraz
   do
      mkdir -p $DATA_ROOT/15_flye/assembly_${name}_min5000/plasmidAssembly_flye/{fasta,blast,RawReads2plasmidassembly_minimap2}
      bedtools bamtofastq -i $DATA_ROOT/15_flye/assembly_${name}_min5000/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_sorted_Scaffolds_unampped.bam -fq $DATA_ROOT/15_flye/assembly_${name}_min5000/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_sorted_Scaffolds_unampped.fastq

      $flye_root -t 7 --pacbio-raw $DATA_ROOT/15_flye/assembly_${name}_min5000/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_sorted_Scaffolds_unampped.fastq --out-dir $DATA_ROOT/15_flye/assembly_${name}_min5000/plasmidAssembly_flye/fasta --genome-size ${size} -m 5000 -i 3 |tee $DATA_ROOT/15_flye/assembly_${name}_min5000/plasmidAssembly_flye/fasta/assemblylog.txt
    size=1m
  done

```
####blast plasmid assembly (after flye)
In the following we blasted both the contigs.fasta aswell as the scaffolds.fasta file against the prokaryotic ncbi database
```{bash blast_plasmid_flye_assemblies ,eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

###-----------------
##blast rawread flye scaffold assembly (entire procaryotic ncbi database)
###-----------------

for name in cottens lapraz
   do
     blastn -num_threads 2 -max_hsps 1 -max_target_seqs 1 -task megablast -show_gis -query $DATA_ROOT/15_flye/assembly_${name}_raw/plasmidAssembly_flye/fasta/scaffolds.fasta -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle slen" -db $blastdb/all_prokaryotes -out $DATA_ROOT/15_flye/assembly_${name}_raw/plasmidAssembly_flye/blast/blastAssemblyaginstScaffolds.txt -evalue 0.01 -word_size 64
   done

###-----------------
##blast min5000 flye assembly (entire procaryotic ncbi database)
###-----------------

for name in cottens lapraz
   do
     blastn -num_threads 2 -max_hsps 1 -max_target_seqs 1 -task megablast -show_gis -query $DATA_ROOT/15_flye/assembly_${name}_min5000/plasmidAssembly_flye/fasta/scaffolds.fasta -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle slen" -db $blastdb/all_prokaryotes -out $DATA_ROOT/15_flye/assembly_${name}_min5000/plasmidAssembly_flye/blast/blastAssemblyaginstScaffolds.txt -evalue 0.01 -word_size 64
   done
```
##plasmid assembly read mapping

```{bash minimapr_plasmidAssemblym, eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

##-------   
##rawRead mappingminimap2 to rawRead Scaffolds assembly
##-------   
for name in cottens lapraz
   do

     ${minimap2_root} -t 12 -ax map-pb $DATA_ROOT/15_flye/assembly_${name}_raw/plasmidAssembly_flye/fasta/scaffolds.fasta $DATA_ROOT/15_flye/assembly_${name}_raw/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_sorted_Scaffolds_unampped.fastq |samtools sort -@8 -O BAM -o $DATA_ROOT/15_flye/assembly_${name}_raw/plasmidAssembly_flye/RawReads2plasmidassembly_minimap2/unmapped_${name}_minimap2Mapped_plasmid_Scaffolds_sorted.bam - |tee $DATA_ROOT/15_flye/assembly_${name}_raw/plasmidAssembly_flye/RawReads2plasmidassembly_minimap2/unmapped_${name}_minimap2Mapped_plasmid_Scaffolds_sorted.log
   samtools -flagstat $DATA_ROOT/15_flye/assembly_${name}_raw/plasmidAssembly_flye/RawReads2plasmidassembly_minimap2/unmapped_${name}_minimap2Mapped_plasmid_Scaffolds_sorted.bam >> $DATA_ROOT/15_flye/assembly_${name}_raw/plasmidAssembly_flye/RawReads2plasmidassembly_minimap2/unmapped_${name}_minimap2Mapped_plasmid_Scaffolds_sorted.log
   done 

##-------   
##rawRead mappingminimap2 to min5000 Scaffolds assembly
##-------   
for name in cottens lapraz
   do

     ${minimap2_root} -t 12 -ax map-pb $DATA_ROOT/15_flye/assembly_${name}_min5000/plasmidAssembly_flye/fasta/scaffolds.fasta $DATA_ROOT/15_flye/assembly_${name}_min5000/rawReads2assembly_minimap2/raw_${name}_minimap2Mapped_sorted_Scaffolds_unampped.fastq |samtools sort -@8 -O BAM -o $DATA_ROOT/15_flye/assembly_${name}_min5000/plasmidAssembly_flye/RawReads2plasmidassembly_minimap2/unmapped_${name}_minimap2Mapped_plasmid_Scaffolds_sorted.bam - |tee $DATA_ROOT/15_flye/assembly_${name}_min5000/plasmidAssembly_flye/RawReads2plasmidassembly_minimap2/unmapped_${name}_minimap2Mapped_plasmid_Scaffolds_sorted.log
   samtools -flagstat $DATA_ROOT/15_flye/assembly_${name}_min5000/plasmidAssembly_flye/RawReads2plasmidassembly_minimap2/unmapped_${name}_minimap2Mapped_plasmid_Scaffolds_sorted.bam >> $DATA_ROOT/15_flye/assembly_${name}_min5000/plasmidAssembly_flye/RawReads2plasmidassembly_minimap2/unmapped_${name}_minimap2Mapped_plasmid_Scaffolds_sorted.log
   done 
```

##Mauve alignment of complete Genomes
In order to see the colinearity of the sofar Sequenced RefSeq genomes and the one sequenced here we implement a Mauve alignement.
For the comparison I downloaded all the completely sequenced reference genomes on RefSeq for Lactobacillus helveticus (n=13), Lactobacillus delbrueckii (n=17) and Streptococcus thermophilus (n=22) on the 29.01.2018. 
For the mauve alignment I split the assembled contigs and aligned them with Mauve according to the best Blast hit. The alignment in Mauve was implemented from the GUI. 

```{bash splitMultiFastA , eval=F}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

##--
##raw reads
##--

for name in cottens lapraz
   do
    location=$(pwd)
    cd $DATA_ROOT/15_flye/assembly_${name}_raw/SmrtLink_polished_reference_arrow/assembly/
    awk '/^>/{s=++d".fasta"} {print > s}' reference_${name}.fasta
    cd $location
  done

##--
##min5000 reads
##--

for name in cottens lapraz
   do
    location=$(pwd)
    cd $DATA_ROOT/15_flye/assembly_${name}_min5000/SmrtLink_polished_reference_arrow/assembly/
    awk '/^>/{s=++d".fasta"} {print > s}' reference_${name}.fasta
    cd $location
  done

```

```{bash progressive_Mauve , eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

###---------------------
##Cottens assemblies mauve
###---------------------

##--------------------------Streptococcus mauve-----------------------------

${mauve_root}/progressiveMauve --output=$DATA_ROOT/19_mauve/cottens_min5000/Ster/all.image.out  \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_900094135.1_ASM90009413v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_002846155.1_ASM284615v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_002846075.1_ASM284607v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_002443035.1_ASM244303v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_002286255.1_ASM228625v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_002012365.1_ASM201236v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_001855705.1_ASM185570v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_001705585.1_ASM170558v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_001685375.1_ASM168537v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_001663795.1_ASM166379v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_001514435.1_ASM151443v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_001280285.1_ASM128028v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_001008015.1_ASM100801v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_000971665.1_ASM97166v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_000698885.1_ASM69888v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_000262675.1_ASM26267v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_000253395.1_ASM25339v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_000182875.1_ASM18287v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_000014485.1_ASM1448v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_000011845.1_ASM1184v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Sthe/fasta/ncbi-genomes-2018-01-29/GCF_000011825.1_ASM1182v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/15_flye/assembly_cottens_min5000/SmrtLink_polished_reference_arrow/assembly/Ster.fasta

##--------------------------Lactobacillus delbrueckii mauve-----------------------------

${mauve_root}/progressiveMauve --output=$DATA_ROOT/19_mauve/cottens_min5000/Ldel/all.image.out  \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Ldel/fasta/ncbi-genomes-2018-01-29/GCF_900196735.1_Lactobacillus_bulgaricus_ACA-DC_87_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Ldel/fasta/ncbi-genomes-2018-01-29/GCF_002285775.1_ASM228577v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Ldel/fasta/ncbi-genomes-2018-01-29/GCF_002278095.1_ASM227809v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Ldel/fasta/ncbi-genomes-2018-01-29/GCF_002142575.1_ASM214257v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Ldel/fasta/ncbi-genomes-2018-01-29/GCF_002000885.1_ASM200088v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Ldel/fasta/ncbi-genomes-2018-01-29/GCF_001953135.1_ASM195313v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Ldel/fasta/ncbi-genomes-2018-01-29/GCF_001908415.1_ASM190841v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Ldel/fasta/ncbi-genomes-2018-01-29/GCF_001888985.1_ASM188898v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Ldel/fasta/ncbi-genomes-2018-01-29/GCF_001888965.1_ASM188896v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Ldel/fasta/ncbi-genomes-2018-01-29/GCF_001888945.1_ASM188894v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Ldel/fasta/ncbi-genomes-2018-01-29/GCF_001888925.1_ASM188892v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Ldel/fasta/ncbi-genomes-2018-01-29/GCF_001888905.1_ASM188890v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Ldel/fasta/ncbi-genomes-2018-01-29/GCF_001469775.1_ASM146977v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Ldel/fasta/ncbi-genomes-2018-01-29/GCF_000191165.1_ASM19116v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Ldel/fasta/ncbi-genomes-2018-01-29/GCF_000182835.1_ASM18283v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Ldel/fasta/ncbi-genomes-2018-01-29/GCF_000056065.1_ASM5606v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/18_NCBIGenomes/Ldel/fasta/ncbi-genomes-2018-01-29/GCF_000014405.1_ASM1440v1_genomic.fna \
    /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/15_flye/assembly_cottens_min5000/SmrtLink_polished_reference_arrow/assembly/Ldel.fasta

###---------------------
##lapraz assemblies mauve
###---------------------



```


#Figtree based on the mauve alignment
Apart from the colinearity plot, mauve additionally outputs a newick file with a prelimary tree. This prelimary tree is visualized in figtree. However before doing that the sequences names have to be match to the orignal file names again. 

```{bash name_matchin , eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

##---
##Ster change name in mauve
##---
sudo rm $DATA_ROOT/19_mauve/cottens_min5000/Ster/all.image.out.guide_tree.renamed
cp $DATA_ROOT/19_mauve/cottens_min5000/Ster/all.image.out.guide_tree $DATA_ROOT/19_mauve/cottens_min5000/Ster/all.image.out.guide_tree.renamed

#for name in cottens lapraz
#   do
    grep -E "Sequence[0-9]+File" $DATA_ROOT/19_mauve/cottens_min5000/Ster/all.image.out |sed 's/#Sequence/seq/g'|sed 's/File//g'|awk -F '[\t/]' '{print $1 , $14}'| sed 's/ /\t/g' |sed 's/.f/_f/g'|sort -r -V > $DATA_ROOT/tmp.txt

    IFS=$'\n'
    while IFS= read -r line; 
      do
        tmp1=$(echo $line | cut -f1)
        tmp2=$(echo $line | cut -f2)
        sed -i "s/${tmp1}/${tmp2}/g" $DATA_ROOT/19_mauve/cottens_min5000/Ster/all.image.out.guide_tree.renamed
      done < "$DATA_ROOT/tmp.txt"
 #     rm $DATA_ROOT/tmp.txt
#   done


##---
##Ldel change name in mauve
##---
sudo rm $DATA_ROOT/19_mauve/cottens_min5000/Ldel/all.img.old.guide_tree.renamed
cp $DATA_ROOT/19_mauve/cottens_min5000/Ldel/all.img.old.guide_tree $DATA_ROOT/19_mauve/cottens_min5000/Ldel/all.img.old.guide_tree.renamed

#for name in cottens lapraz
#   do
    grep -E "Sequence[0-9]+File" $DATA_ROOT/19_mauve/cottens_min5000/Ldel/all.img.old |sed 's/#Sequence/seq/g'|sed 's/File//g'|awk -F '[\t/]' '{print $1 , $14}'| sed 's/ /\t/g' |sed 's/.f/_f/g'|sort -r -V > $DATA_ROOT/tmp.txt

    IFS=$'\n'
    while IFS= read -r line; 
      do
        tmp1=$(echo $line | cut -f1)
        tmp2=$(echo $line | cut -f2)
        sed -i "s/${tmp1}/${tmp2}/g" $DATA_ROOT/19_mauve/cottens_min5000/Ldel/all.img.old.guide_tree.renamed
      done < "$DATA_ROOT/tmp.txt"
 #     rm $DATA_ROOT/tmp.txt
#   done

```
##Figtree
With Figtree I visualized the preliminary tree outputed from Mauve.
The Cottens trees look as following:
![Lactobacillus delbrueckii tree containing the here assembled (Ldel_fasta) and all complete RefSeqs of Ldel](/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/19_mauve/cottens_min5000/Ldel/all.img.old.guide_tree.renamed_tree.png)

![Lactobacillus delbrueckii tree containing the here assembled (Ldel_fasta) and all complete RefSeqs of Ldel](/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/19_mauve/cottens_min5000/Ster/all.image.out.guide_tree.copy_tree.png)



##Abruijn
Abruijn is an assembler from the Pevzner lab that has shown to be very efficient and easy to use with different kinds of sequencing date. 
```{bash abruijn , warning=F, eval=F}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

##---------------
##extract raw data
##---------------
for name in cottens lapraz
   do
     tar -zxvf $DATA_ROOT/01_SMRTlink_data/${name}/reads.tar.gz -C $DATA_ROOT/01_SMRTlink_data/${name}/
   done

date=$(date +%Y%m%d)
##---------------
##creat abruijn directories
##---------------
for name in cottens lapraz
   do
    mkdir -p $DATA_ROOT/05_abruijn/${date}assembly_${name}/{quast,fasta}
   done
##---------------
##abruijn assembly with ALL subreads
##---------------
cov=200
for name in cottens lapraz
  do
    $abruijn_root/abruijn -t 16  $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fastq $DATA_ROOT/05_abruijn/assembly_${name}/fasta ${cov} -p pacbio -o 2500
    cov=900
  done

for name in cottens lapraz
   do
    quast.py $DATA_ROOT/05_abruijn/assembly_${name}/fasta/polished_1.fasta -t 8 -o $DATA_ROOT/05_abruijn/assembly_${name}/quast/
   done


##---------------
##abruijn assembly with subreads > 5000bp
##---------------
for name in cottens lapraz
  do
    prinseq-lite.pl -fastq $DATA_ROOT/01_SMRTlink_data/${name}/reads.fastq -min_len 7000 -out_format 3 -out_good $DATA_ROOT/01_SMRTlink_data/${name}/reads_min5000bp  -out_bad null
  done
  
cov=100
for name in cottens lapraz
  do
    mkdir -p $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/{fasta,quast}
    $abruijn_root/abruijn -t 10 $DATA_ROOT/01_SMRTlink_data/${name}/reads_min5000bp.fastq $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/fasta ${cov} -p pacbio -o 2500
    cov=300
  done
##quast
for name in cottens lapraz
   do
    quast.py $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/fasta/polished_1.fasta -t 8 -o $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/quast/
   done
   
##blast
date=20171114
for name in cottens lapraz
   do
     mkdir -p $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/blast
     $blast_root/blastn -num_threads 16 -max_hsps 1 -max_target_seqs 1 -task megablast -show_gis -query $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/fasta/polished_1.fasta -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle slen" -db $blastdb/all_prokaryotes -out $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/blast/blastAssembly.txt -evalue 0.01 -word_size 64
   done   
##---------------
##abruijn assembly with polished reads (preassembly pacbio)
##---------------

  
cov=100
for name in cottens lapraz
  do
    mkdir -p $DATA_ROOT/05_abruijn/assembly_${name}_preassembledReads/{fasta,quast}
    $abruijn_root/abruijn -t 10 01_SMRTlink_data/${name}/preassembled_reads/preads4falcon.fasta $DATA_ROOT/05_abruijn/assembly_${name}_preassembledReads/fasta ${cov} -p pacbio -i 2 -o 2500
    cov=300
  done
##quast
for name in cottens lapraz
   do
    quast.py $DATA_ROOT/05_abruijn/assembly_${name}_preassembledReads/fasta/polished_1.fasta -t 8 -o $DATA_ROOT/05_abruijn/assembly_${name}_preassembledReads/quast/
   done
##blast
date=20171114
for name in cottens lapraz
   do
     mkdir -p $DATA_ROOT/05_abruijn/assembly_${name}_preassembledReads/blast
     $blast_root/blastn -num_threads 16 -max_hsps 1 -max_target_seqs 1 -task megablast -show_gis -query $DATA_ROOT/05_abruijn/assembly_${name}_preassembledReads/fasta/polished_1.fasta -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle slen" -db $blastdb/all_prokaryotes -out $DATA_ROOT/05_abruijn/assembly_${name}_preassembledReads/blast/blastAssembly.txt -evalue 0.01 -word_size 64
   done
   
###==========================
##coverage
###==========================  
for name in cottens lapraz
   do
    mkdir -p $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/coverage  
    $bbmap_root/bbwrap.sh ref=$DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/fasta/polished_1.fasta 
        $bbmap_root/bbwrap.sh ref=$DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/fasta/polished_1.fasta nodisk in=10 $DATA_ROOT/01_SMRTlink_data/${name}/reads_min5000bp.fastq out=$DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/coverage 
  done


```
##Look at assembled contigs
In the following we look at the assembled contigs



```{bash assembly_check ,eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

for name in cottens lapraz
   do
#    numCon=${grep ">" $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/fasta/polished_1.fasta}
    echo "The assembly of" ${name}:
    cat $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/blast/blastAssembly.txt
    
   done

```


```{r assembly_check_inR ,message=FALSE ,warning=FALSE,eval=FALSE}
##-----------
##import files
##-----------
assembly_cottens <- read_delim("~/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/05_abruijn/assembly_cottens_min5000bp/blast/blastAssembly.txt", "\t", escape_double = FALSE, col_names = c("qseqid" ,"sseqid" ,"pident" ,"length" ,"mismatch", "gapopen", "qstart", "qend", "sstart", "send","evalue" ,"bitscore" ,"stitle"), trim_ws = TRUE)

assembly_lapraz <- read_delim("~/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/05_abruijn/assembly_lapraz_min5000bp/blast/blastAssembly.txt", "\t", escape_double = FALSE, col_names = c("qseqid" ,"sseqid" ,"pident" ,"length" ,"mismatch", "gapopen", "qstart", "qend", "sstart", "send","evalue" ,"bitscore" ,"stitle"), trim_ws = TRUE)

##-----------
##add meta information
##-----------
for(i in c("cottens","lapraz")){
    nam <- paste("assembly_",i,sep="")
    name <-   data.frame(do.call('rbind',strsplit(get(nam)$qseqid, split = "_")))
    colnames(name) <- c("circularity","number","contigLength","ContigCoverage")
    assign(nam,cbind(get(nam),name))
}
##-----------
##add meta information
##-----------

for(i in c("cottens","lapraz")){
    nam <- paste("assembly_",i,sep="")
    nam_2 <- paste("reduced_assembly_",i,sep="")
    nam_2 <- as.data.frame(get(nam)) %>%
    dplyr::select(qseqid, circularity, contigLength, ContigCoverage,bitscore,sseqid,stitle)
    print(c("The Best hits for the ",i, "assembly are the following"))
    kable(nam_2)
}




```


##Minimap2 and Miniasm from Heng Li
The recently developed Minimap2 (Mapper) amd Miniasm (overlaper) are much faster as all other methods. 
However it did not improve the Metagenomic assembly. Therefor the chunk was removed on the 24.1.2018. The files however still exist.


##Metaspades and megahit as a Debruijn assembler specifically for metagenomic datasets
Although Debruijn assemblers are orignally meant for short and low error rate sequences we give it a try on the reads >5000bp.

Nor the Metaspades aswell as the megahit assembly did not work for the pacbio only data.
Therefor it was removed on the 24.1.2018

##sprai error correction
Sprai is able to do the error correction without assembly.
Additionally spray can check for circularity of an assembly.
However this code chunk was removed on the 24.1.2018.

#Binning reads/preassembled reads 
Different binning approachs have been applied. Amongst the most common ones are: the Tetranucleotide binning [@Iverson2012] and the coverage based binning [@Albertsen2013]. However it has been noticed that both entities applie only to entities with a extreme skew. Further it is unlikely that both the coverage aswell as the tetranucleotide distribution is consisten over the entire genome. Therefore these methods might additionally bias the assembly. 

##Own Blobology on preassemblies
The blobolgoy is based on te preassembled reads.
We need the following data:
1. Preassembled reads coverage (#RawReads minimap2 aligned to preassembled reads)
2. GC% of preassembled reads
3. Minimap2 aligning of preassembled reads to Flye assembly

After this we make a plot in R. 

```{bash blobology_preassembly ,eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

###==========================
##1. Preassembled reads coverage 
###==========================
##-------   
##index reference genome
##-------   

for name in cottens lapraz
   do
    $minimap2_root -d $DATA_ROOT/01_SMRTlink_data/${name}/preassembled_reads/preads4falcon.mmi $DATA_ROOT/01_SMRTlink_data/${name}/preassembled_reads/preads4falcon.fasta
   done

##-------   
##read mapping
##-------   
for name in cottens lapraz
   do
     mkdir -p $DATA_ROOT/16_blobologyOwn/preassembly_${name}/{01_minimapRawReads,02_GC,03_minimapAgainstFlye}
     mkdir -p $DATA_ROOT/16_blobologyOwn/preassembly_${name}/01_minimapRawReads/bamqc
 
     ${minimap2_root} -ax map-pb $DATA_ROOT/01_SMRTlink_data/${name}/preassembled_reads/preads4falcon.fasta $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fastq |samtools sort -@8 -O BAM -o $DATA_ROOT/16_blobologyOwn/preassembly_${name}/01_minimapRawReads/raw_${name}_mapped_preassembly_sorted.bam - | tee $DATA_ROOT/16_blobologyOwn/preassembly_${name}/01_minimapRawReads/raw_${name}_mapped_preassembly_sorted.log
     
      $QUAL_ROOT/qualimap bamqc -bam $DATA_ROOT/16_blobologyOwn/preassembly_${name}/01_minimapRawReads/raw_${name}_mapped_preassembly_sorted.bam -outdir $DATA_ROOT/16_blobologyOwn/preassembly_${name}/01_minimapRawReads/bamqc/ --java-mem-size=2G
     
   done 

for name in cottens lapraz
   do 
    samtools view $DATA_ROOT/16_blobologyOwn/preassembly_${name}/01_minimapRawReads/raw_${name}_mapped_preassembly_sorted.bam |cut -f 3 |sort |uniq -c |awk '{$1=$1};1'| sed -e 's/ /\t/g' > $DATA_ROOT/16_blobologyOwn/preassembly_${name}/01_minimapRawReads/coverage_${name}_preassembled.txt
   done

###==========================
##2. GC% of preassembled reads 
###==========================
for name in cottens lapraz
   do
    infoseq -nocolumns -delimiter , -auto -only -name -length -pgc $DATA_ROOT/01_SMRTlink_data/${name}/preassembled_reads/preads4falcon.fasta > $DATA_ROOT/16_blobologyOwn/preassembly_${name}/02_GC/gc_content_${name}_preassemeled.txt
   
    #prinseq-lite.pl -stats_dinuc -fasta $DATA_ROOT/01_SMRTlink_data/${name}/preassembled_reads/preads4falcon.fasta > $DATA_ROOT/16_blobologyOwn/preassembly_${name}/02_GC/gc_content_${name}_preassemeled.txt
  done

###==========================
##3. Minimap2 aligning of preassembled reads to Flye assembly
###==========================
for name in cottens lapraz
   do
     mkdir -p $DATA_ROOT/16_blobologyOwn/preassembly_${name}/03_minimapAgainstFlye/bamqc

    ${minimap2_root} -ax map-pb $DATA_ROOT/15_flye/assembly_${name}_min5000/fasta/scaffolds.fasta $DATA_ROOT/01_SMRTlink_data/${name}/preassembled_reads/preads4falcon.fasta |samtools sort -@8 -O BAM -o $DATA_ROOT/16_blobologyOwn/preassembly_${name}/03_minimapAgainstFlye/preassembled_${name}_mapped_FlyeAssembly_sorted.bam - | tee $DATA_ROOT/16_blobologyOwn/preassembly_${name}/03_minimapAgainstFlye/preassembled_${name}_mapped_FlyeAssembly_sorted.log
    
    $QUAL_ROOT/qualimap bamqc -bam $DATA_ROOT/16_blobologyOwn/preassembly_${name}/03_minimapAgainstFlye/preassembled_${name}_mapped_FlyeAssembly_sorted.bam -outdir $DATA_ROOT/16_blobologyOwn/preassembly_${name}/03_minimapAgainstFlye/bamqc/ --java-mem-size=2G
    
    done

for name in cottens lapraz
   do 
    samtools view $DATA_ROOT/16_blobologyOwn/preassembly_${name}/03_minimapAgainstFlye/preassembled_${name}_mapped_FlyeAssembly_sorted.bam |awk '{print $1 , $3}'  > $DATA_ROOT/16_blobologyOwn/preassembly_${name}/03_minimapAgainstFlye/mapping_${name}_preassembled.txt
   done
   

```

```{r analysis_blobology,eval=TRUE,warning=FALSE}
##preassembly coverage
coverage_preassembly_cottens <- read_delim("/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/16_blobologyOwn/preassembly_cottens/01_minimapRawReads/coverage_cottens_preassembled.txt", "\t", escape_double = FALSE, col_names = c("hits","name"),  trim_ws = TRUE)
coverage_preassembly_cottens <- coverage_preassembly_cottens[-1,]
coverage_preassembly_lapraz <- read_delim("/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/16_blobologyOwn/preassembly_lapraz/01_minimapRawReads/coverage_lapraz_preassembled.txt", "\t", escape_double = FALSE, col_names = c("hits","name"),  trim_ws = TRUE)
coverage_preassembly_lapraz <- coverage_preassembly_lapraz[-1,]

##gc content 
gc_cottens <- read_csv("/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/16_blobologyOwn/preassembly_cottens/02_GC/gc_content_cottens_preassemeled.txt")
gc_lapraz <- read_csv("/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/16_blobologyOwn/preassembly_lapraz/02_GC/gc_content_lapraz_preassemeled.txt")  

##reference contig 

mapping_cottens_preassembled <- read_delim("/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/16_blobologyOwn/preassembly_cottens/03_minimapAgainstFlye/mapping_cottens_preassembled.txt",  " ", escape_double = FALSE, col_names = c("read","scaffold"),trim_ws = TRUE)
mapping_lapraz_preassembled <- read_delim("/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/16_blobologyOwn/preassembly_lapraz/03_minimapAgainstFlye/mapping_lapraz_preassembled.txt",  " ", escape_double = FALSE, col_names = c("read","scaffold"),trim_ws = TRUE) 

##ordering contigs Cottens
all(mapping_cottens_preassembled$read %in% coverage_preassembly_cottens$name)
all(mapping_cottens_preassembled$read %in% gc_cottens$Name)

no_cov <- nrow(mapping_cottens_preassembled)-nrow(coverage_preassembly_cottens)
print(c(no_cov,"contigs do not have a coverage"))

gc_cottens <- gc_cottens[match(coverage_preassembly_cottens$name,gc_cottens$Name),]
mapping_cottens_preassembled <- mapping_cottens_preassembled[match(coverage_preassembly_cottens$name,mapping_cottens_preassembled$read),]

all(mapping_cottens_preassembled$read %in% coverage_preassembly_cottens$name)
all(mapping_cottens_preassembled$read %in% gc_cottens$Name)

left_contigs <- nrow(mapping_cottens_preassembled)
print(c("Finally, there are ",left_contigs,"contigs left"))


##ordering contigs lapraz

all(mapping_lapraz_preassembled$read %in% coverage_preassembly_lapraz$name)
all(mapping_lapraz_preassembled$read %in% gc_lapraz$Name)

no_cov <- nrow(mapping_lapraz_preassembled)-nrow(coverage_preassembly_lapraz)
print(c(no_cov,"contigs do not have a coverage"))

gc_lapraz <- gc_lapraz[match(coverage_preassembly_lapraz$name,gc_lapraz$Name),]
mapping_lapraz_preassembled <- mapping_lapraz_preassembled[match(coverage_preassembly_lapraz$name,mapping_lapraz_preassembled$read),]

all(mapping_lapraz_preassembled$read %in% coverage_preassembly_lapraz$name)
all(mapping_lapraz_preassembled$read %in% gc_lapraz$Name)

left_contigs <- nrow(mapping_lapraz_preassembled)
print(c("Finally, there are ",left_contigs,"contigs left"))

##creat dataframe cottens

final_df_cottens <- data.frame(name=mapping_cottens_preassembled$read,coverage=coverage_preassembly_cottens$hits,gc=gc_cottens$`%GC`,length=gc_cottens$Length,mapping=mapping_cottens_preassembled$scaffold)

##creat dataframe lapraz

final_df_lapraz <- data.frame(name=mapping_lapraz_preassembled$read,coverage=coverage_preassembly_lapraz$hits,gc=gc_lapraz$`%GC`,length=gc_lapraz$Length,mapping=mapping_lapraz_preassembled$scaffold)

##plot cottens

scatterPlot <- ggplot(data = final_df_cottens,aes(x=coverage,y=gc,color=mapping))+
  geom_point()+
  theme_bw()+
  theme(legend.title=element_blank())

# Marginal density plot of x (top panel)
xdensity <- ggplot(data = final_df_cottens,aes(x=coverage,fill=mapping)) + 
  geom_density(alpha=.5) + 
  scale_x_continuous(name="") +
  theme_bw()+
  theme(legend.position="none",axis.text.x=element_blank())
# xdensity
# Marginal density plot of y (right panel)
ydensity <- ggplot(data = final_df_cottens, aes(x=gc, fill=mapping)) + 
  geom_density(alpha=.5) + 
  scale_x_continuous(name="")+
  theme_bw()+
  coord_flip()+
  theme(legend.position="none",axis.text.x=element_blank())
# ydensity

legend <- g_legend(scatterPlot)

grid.arrange(xdensity, legend,scatterPlot+theme(legend.position="none"), ydensity, 
        ncol=2, nrow=2, widths=c(4, 1.4), heights=c(1.4, 4))

##plot lapraz
# 
# scatterPlot <- ggplot(data = final_df_lapraz,aes(x=coverage,y=gc,color=mapping))+
#   geom_point()+
#   theme_bw()+
#   theme(legend.title=element_blank())
# 
# # Marginal density plot of x (top panel)
# xdensity <- ggplot(data = final_df_lapraz,aes(x=coverage,fill=mapping)) + 
#   geom_density(alpha=.5) + 
#   scale_x_continuous(name="") +
#   theme_bw()+
#   theme(legend.position="none",axis.text.x=element_blank())
# # xdensity
# # Marginal density plot of y (right panel)
# ydensity <- ggplot(data = final_df_lapraz, aes(x=gc, fill=mapping)) + 
#   geom_density(alpha=.5) + 
#   scale_x_continuous(name="")+
#   theme_bw()+
#   coord_flip()+
#   theme(legend.position="none",axis.text.x=element_blank())
# # ydensity
# 
# legend <- g_legend(scatterPlot)
# 
# grid.arrange(xdensity, legend,scatterPlot+theme(legend.position="none"), ydensity, 
#         ncol=2, nrow=2, widths=c(4, 1.4), heights=c(1.4, 4))

```
The preassembled reads divide very well between the different contigs according to the GC. However there are certain preassembled reads which are intermediat or do not follow this pattern at all. Therefore binning the preassembled reads before the assembly would not make sense and most likely bias the assembly process.  

##Nucmer to align different assemblies against eachother

The large limitation in Assembling genomes as well as metagenomes is the ability to sequence larger than repeats. For a single genome this means the read length needs to be sufficiently long to span the largest repeat. However for metagenome assembly there is additonally to the intragenomic repeat a difficulty to dinstinguish inter genomic repeats [@Olson2017]. In order to look at the repeats

```{bash nucmer ,eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

###==========================
##combine delbruecki and streptococcus from lapraz abruijn assembly
###==========================

mkdir -p $DATA_ROOT/11_nucmer_repeat/cottens_abruijn_delbruecki_streptococcus/{fasta,nucmer,deltaFiltered}

cat $DATA_ROOT/05_abruijn/assembly_cottens_min5000bp/fasta/multifasta/linear_1_len1924117_cov316.fa $DATA_ROOT/05_abruijn/assembly_cottens_min5000bp/fasta/multifasta/linear_2_len2265372_cov167.fa > $DATA_ROOT/11_nucmer_repeat/cottens_abruijn_delbruecki_streptococcus/fasta/reads.fa

###==========================
##nucmer
###==========================
cd $DATA_ROOT/11_nucmer_repeat/cottens_abruijn_delbruecki_streptococcus/nucmer/
nucmer -maxmatch -nosimplify $DATA_ROOT/11_nucmer_repeat/cottens_abruijn_delbruecki_streptococcus/fasta/reads.fa $DATA_ROOT/11_nucmer_repeat/cottens_abruijn_delbruecki_streptococcus/fasta/reads.fa 
cd $DATA_ROOT

###==========================
##delta filter
###==========================
mkdir -p $DATA_ROOT/11_nucmer_repeat/cottens_abruijn_delbruecki_streptococcus/{fasta,nucmer,deltaFiltered}
delta-filter -i 85 -l 3000  $DATA_ROOT/11_nucmer_repeat/cottens_abruijn_delbruecki_streptococcus/nucmer/out.delta > $DATA_ROOT/11_nucmer_repeat/cottens_abruijn_delbruecki_streptococcus/deltaFiltered/filtered.txt
##split files
cd $DATA_ROOT/11_nucmer_repeat/cottens_abruijn_delbruecki_streptococcus/deltaFiltered/
split -t ">" -l 1 $DATA_ROOT/11_nucmer_repeat/cottens_abruijn_delbruecki_streptococcus/deltaFiltered/filtered.txt
cd $DATA_ROOT
##extract only necassary files
grep "[[:space:]]" $DATA_ROOT/11_nucmer_repeat/cottens_abruijn_delbruecki_streptococcus/deltaFiltered/xac | awk 'BEGIN {OFS ="\t"} {print "Sterm" , $1 , $2 , "Ldel" , $3 , $4}' > $DATA_ROOT/11_nucmer_repeat/cottens_abruijn_delbruecki_streptococcus/deltaFiltered/links.txt
tail -1 $DATA_ROOT/11_nucmer_repeat/cottens_abruijn_delbruecki_streptococcus/deltaFiltered/links.txt > $DATA_ROOT/11_nucmer_repeat/cottens_abruijn_delbruecki_streptococcus/deltaFiltered/link.txt
rm $DATA_ROOT/11_nucmer_repeat/cottens_abruijn_delbruecki_streptococcus/deltaFiltered/links.txt

###==========================
##Lhev and Sterm repeats
###==========================
mkdir -p $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/{01_fasta,nucmer,deltaFiltered}
cat $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/01_fasta/Lhev.fasta $DATA_ROOT/05_abruijn/assembly_cottens_min5000bp/fasta/multifasta/linear_1_len1924117_cov316.fa > $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/01_fasta/Lhev_Sterm.fasta
##nucmer
cd $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/nucmer/
nucmer -maxmatch -nosimplify $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/01_fasta/Lhev_Sterm.fasta $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/01_fasta/Lhev_Sterm.fasta
cd $DATA_ROOT
##delta-filter
delta-filter -i 85 -l 3000  $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/nucmer/out.delta >  $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/deltaFiltered/filtered.txt
##split files
cd $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm//deltaFiltered/
split -t ">" -l 1 $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/deltaFiltered/filtered.txt
cd $DATA_ROOT
##extract only necassary files
grep "[[:space:]]" $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/deltaFiltered/xac | awk 'BEGIN {OFS ="\t"} {print "Sterm" , $1 , $2 , "Lhev" , $3 , $4}' > $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/deltaFiltered/links.txt
tail -1 $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/deltaFiltered/links.txt > $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/deltaFiltered/link.txt
rm $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/deltaFiltered/links.txt
##size of repeats
awk '{print $3-$2}' $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/deltaFiltered/link.txt


###==========================
##Lhev and delbruecki repeats
###==========================
mkdir -p $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/{01_fasta,nucmer,deltaFiltered}
cat $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/01_fasta/Lhev.fasta $DATA_ROOT/05_abruijn/assembly_cottens_min5000bp/fasta/multifasta/linear_2_len2265372_cov167.fa > $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/01_fasta/Lhev_del.fasta
##nucmer
cd $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/nucmer/
nucmer -maxmatch -nosimplify $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/01_fasta/Lhev_del.fasta $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/01_fasta/Lhev_del.fasta
cd $DATA_ROOT
##delta-filter
delta-filter -i 85 -l 3000  $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/nucmer/out.delta >  $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/deltaFiltered/filtered.txt
##split files
cd $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/deltaFiltered/
split -t ">" -l 1 $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/deltaFiltered/filtered.txt
cd $DATA_ROOT
##extract only necassary files
grep "[[:space:]]" $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/deltaFiltered/xac | awk 'BEGIN {OFS ="\t"} {print "Ldel" , $1 , $2 , "Lhev" , $3 , $4}' > $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/deltaFiltered/links.txt
tail -1 $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/deltaFiltered/links.txt > $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/deltaFiltered/link.txt
rm $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/deltaFiltered/links.txt

###==========================
##cat all three intergenomic repeat files together
###==========================

cat $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/deltaFiltered/link.txt $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/deltaFiltered/link.txt $DATA_ROOT/11_nucmer_repeat/cottens_abruijn_delbruecki_streptococcus/deltaFiltered/link.txt > $DATA_ROOT/11_nucmer_repeat/links.txt


###==========================
##Lhev and Lacido repeats
###==========================
mkdir -p $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/{01_fasta,nucmer,deltaFiltered}
cat $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/01_fasta/Lhev.fasta $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/01_fasta/Lacido.fasta > $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/01_fasta/Lhev_Lacido.fasta
##nucmer
cd $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/nucmer/
nucmer -maxmatch -nosimplify $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/01_fasta/Lhev_Lacido.fasta $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/01_fasta/Lhev_Lacido.fasta
cd $DATA_ROOT
##delta-filter
delta-filter -i 85 -l 3000  $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/nucmer/out.delta >  $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/deltaFiltered/filtered.txt
##split files
cd $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/deltaFiltered/
split -t ">" -l 1 $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/deltaFiltered/filtered.txt
cd $DATA_ROOT
##extract only necassary files
grep "[[:space:]]" $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/deltaFiltered/xac | awk 'BEGIN {OFS ="\t"} {print "Lhev" , $1 , $2 , "Lacido" , $3 , $4}' > $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/deltaFiltered/links.txt
tail -18 $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/deltaFiltered/links.txt > $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/deltaFiltered/link.txt
rm $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/deltaFiltered/links.txt

###==========================
##Lhev repeats
###==========================

##extract only necassary files
grep "[[:space:]]" $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/deltaFiltered/xab | awk 'BEGIN {OFS ="\t"} {print "Lhev" , $1 , $2 , "Lhev" , $3 , $4}' > $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/deltaFiltered/links_intra.txt
tail -2 $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/deltaFiltered/links_intra.txt > $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/deltaFiltered/link_intra.txt
rm $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/deltaFiltered/links_intra.txt

###==========================
##Ldel repeats
###==========================

##extract only necassary files
grep "[[:space:]]" $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/deltaFiltered/xab | awk 'BEGIN {OFS ="\t"} {print "Ldel" , $1 , $2 , "Ldel" , $3 , $4}' > $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/deltaFiltered/links_intra.txt
tail -18 $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/deltaFiltered/links_intra.txt > $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/deltaFiltered/link_intra.txt
rm $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/deltaFiltered/links_intra.txt


###==========================
##Sterm repeats
###==========================

grep "[[:space:]]" $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/deltaFiltered/xab | awk 'BEGIN {OFS ="\t"} {print "Sterm" , $1 , $2 , "Sterm" , $3 , $4}' > $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/deltaFiltered/links_intra.txt
tail -18 $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/deltaFiltered/links_intra.txt > $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/deltaFiltered/link_intra.txt
rm $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/deltaFiltered/links_intra.txt
##size of repeats

###==========================
##cat all three intragenomic repeat files together
###==========================

cat $DATA_ROOT/11_nucmer_repeat/delbruecki_helveticus/deltaFiltered/link_intra.txt $DATA_ROOT/11_nucmer_repeat/Lhev_Sterm/deltaFiltered/link_intra.txt $DATA_ROOT/11_nucmer_repeat/Lhev_Lacido/deltaFiltered/link_intra.txt > $DATA_ROOT/11_nucmer_repeat/link_intra.txt


```


##Blasting the assemblages against the entire prokaryotic complete genome database
After having assembled the genomes with different assembly algorithms we continue to blast all contigs against the current prokaryotic complete genome database.
In the following we creat the complete genome database and locally BLAST against it.


```{bash BLAST , warning=F, eval=F}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

###-----------------
##create complete prokaryotic genomes blast database
###-----------------
ls $blastdb | awk -F. '{print $1, $2}' | sed 's/ /./g'> $blastdb/filelist.txt

for FILE in `cat $blastdb/filelist.txt`
  do
    gunzip -c $blastdb/${FILE}*
    perl $DATA_ROOT/02_scripts/gbk2fasta.pl < $blastdb/${FILE}* >> $blastdb/all_prokaryotes.fa
    rm $blastdb/${FILE}*.gbff
  done


##creat blastdb
$blast_root/makeblastdb -max_file_sz 10GB -in $blastdb/all_prokaryotes.fa -out $blastdb/all_prokaryotes -parse_seqids -dbtype nucl
###-----------------
##blast Hgap4 assembly
###-----------------

date=20171114
for name in cottens lapraz
   do
     mkdir -p $DATA_ROOT/04_HGAP4/${date}assembly_${name}/blast/
     $blast_root/blastn -num_threads 16 -max_hsps 2 -max_target_seqs 2 -task megablast -show_gis -query $DATA_ROOT/04_HGAP4/${date}assembly_${name}/fasta/assembly.fasta -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle slen" -db $blastdb/all_prokaryotes -out $DATA_ROOT/04_HGAP4/${date}assembly_${name}/blast/blastAssembly_04.txt -evalue 0.01 -word_size 64
   done
##sort for best hit
#sort -k1,1 -k12,12nr -k11,11n  $DATA_ROOT/04_HGAP4/20171114assembly_lapraz/blast/blastAssembly_03.txt | sort -u -k1,1 --merge

date=20171114
for name in cottens lapraz
   do
      echo species level for ${name}
           sort -k1,1 -k12,12nr -k11,11n $DATA_ROOT/04_HGAP4/${date}assembly_${name}/blast/blastAssembly_04.txt | sort -u -k1,1 --merge|cut -f 13|awk 'BEGIN {OFS=" "}; {print $1, $2}'|sort|uniq -c
      echo strain level for ${name}
      sort -k1,1 -k12,12nr -k11,11n  $DATA_ROOT/04_HGAP4/${date}assembly_${name}/blast/blastAssembly_04.txt | sort -u -k1,1 --merge|cut -f 13|sort| uniq -c
   done

###-----------------
##blast ABruijn assembly
###-----------------

date=20171114
for name in cottens lapraz
   do
     mkdir -p $DATA_ROOT/05_abruijn/assembly_${name}/blast/
     $blast_root/blastn -num_threads 16 -max_hsps 2 -max_target_seqs 2 -task megablast -show_gis -query   $DATA_ROOT/05_abruijn/assembly_${name}/fasta/polished_1.fasta -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle slen" -db $blastdb/all_prokaryotes -out $DATA_ROOT/05_abruijn/assembly_${name}/blast/blastAssembly.txt -evalue 0.01 -word_size 64
   done
##sort for best hit
#sort -k1,1 -k12,12nr -k11,11n  $DATA_ROOT/05_abruijn/assembly_lapraz/blast/blastAssembly.txt | sort -u -k1,1 --merge
##uniq strains
#sort -k1,1 -k12,12nr -k11,11n  $DATA_ROOT/05_abruijn/assembly_lapraz/blast/blastAssembly.txt | sort -u -k1,1 --merge|cut -f 13|sort| uniq -c
##species

date=20171114
for name in cottens lapraz
   do
      echo species level for ${name}
           sort -k1,1 -k12,12nr -k11,11n  $DATA_ROOT/05_abruijn/assembly_${name}/blast/blastAssembly.txt | sort -u -k1,1 --merge|cut -f 13|awk 'BEGIN {OFS=" "}; {print $1, $2}'|sort|uniq -c
      echo strain level for ${name}
      sort -k1,1 -k12,12nr -k11,11n  $DATA_ROOT/05_abruijn/assembly_lapraz/blast/blastAssembly.txt | sort -u -k1,1 --merge|cut -f 13|sort| uniq -c
   done




```

## Align raw reads to assembled genomes
I a proceeding step we create a BLAST database of the assembled metagenomic contigs and aligning all raw reads against them. Thereby we check how complete the current metagenome assembly is. 

```{bash minimap_rawreads_assembly, eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh
##=============
##minimap2
##=============

##-------   
##index reference genome
##-------   

#for name in cottens lapraz
#   do
#    $minimap2_root -d $DATA_ROOT/05_abruijn/assembly_${name}/fasta/polished_1.mmi $DATA_ROOT/05_abruijn/assembly_${name}/fasta/polished_1.fasta
#   done

##-------   
##read mapping
##-------   
for name in cottens lapraz
   do
     mkdir -p $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/
     
     ${minimap2_root} -ax map-pb $DATA_ROOT/05_abruijn/assembly_${name}/fasta/polished_1.fasta $DATA_ROOT/01_SMRTlink_data/${name}/reads_${name}_raw.fastq > $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/raw_${name}_mapped_genome.sam
     
   done 
```


```{bash raw_read_align , warning=F, eval=F}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

##=============
##Abruijn
##=============
##----------------
##creat abruijn bwa index
##----------------
date=20171114
for name in cottens lapraz
   do
    bwa index $DATA_ROOT/05_abruijn/assembly_${name}/fasta/polished_1.fasta 
  done
###-----------------
##align raw reads against abruijn assembly
###-----------------
date=20171114
for name in cottens lapraz
   do
     mkdir -p $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/
     bwa mem $DATA_ROOT/05_abruijn/assembly_${name}/fasta/polished_1.fasta $DATA_ROOT/01_SMRTlink_data/${name}/reads.fastq -x pacbio -t 12|samtools sort -@8 -O BAM -o $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/aligned_sorted.bam -
   done
   
###-----------------
##align raw reads against abruijn >7000bp  assembly
###-----------------
##index
date=20171114
for name in cottens lapraz
   do
    bwa index $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/fasta/polished_1.fasta 
  done
date=20171114
for name in cottens lapraz
   do
     mkdir -p $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/reads2assembly_align/
     bwa mem $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/fasta/polished_1.fasta $DATA_ROOT/01_SMRTlink_data/${name}/reads.fastq -x pacbio -t 6|samtools sort -@8 -O BAM -o$DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/reads2assembly_align/aligned_sorted.bam -
   done   



##=============
##Hgap4
##=============
##----------------
##creat hagp2 bwa index
##----------------
date=20171114
for name in cottens lapraz
   do
    bwa index $DATA_ROOT/04_HGAP4/${date}assembly_${name}/fasta/assembly.fasta
  done
###-----------------
##align raw reads against hgap4 assembly
###-----------------
date=20171114
for name in cottens lapraz
   do
     mkdir -p $DATA_ROOT/04_HGAP4/${date}assembly_${name}/reads2assembly_align/
     bwa mem $DATA_ROOT/04_HGAP4/${date}assembly_${name}/fasta/assembly.fasta $DATA_ROOT/01_SMRTlink_data/${name}/reads.fastq -x pacbio -t 12|samtools sort -@8 -O BAM -o $DATA_ROOT/04_HGAP4/${date}assembly_${name}/reads2assembly_align/aligned_sorted.bam -
   done    

```
Aligning with a long noisy read specialized aligner is more stringent. Therfore we prefere the usage of Minimap2 over bwa.

##Divide well matching reads and not aligning reads
To see the completness of the metagenome assembly and potential additional microbial sequences we divide the aligning and the non-aligning reads.
We do this based on the MAPQ score. 
```{bash unmapped_mapped_READS , warning=F, eval=F}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

##=============
##Abruijn
##=============
date=
#date=20171114
for name in cottens_min5000bp lapraz_min5000bp
   do
      mkdir -p $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/bamqc
      samtools view -F 4 $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/aligned_sorted.bam > $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/aligned_sorted_mapped.bam 
      samtools view -b -f 4 $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/aligned_sorted.bam > $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/aligned_sorted_unmapped.bam
      bedtools bamtofastq -i $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/aligned_sorted_unmapped.bam -fq $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/aligned_sorted_unmapped.fastq
      $QUAL_ROOT/qualimap bamqc -bam $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/aligned_sorted.bam -outdir $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/bamqc/ --java-mem-size=2G
      grep "Globals" -A 12  $DATA_ROOT/05_abruijn/20171114assembly_${name}/reads2assembly_align/bamqc/genome_results.txt |tail -11|sed '/^\s*$/d'|sed 's/=/\t/g'|sed -e 's/(.*%)//'|sed 's/,//g' > $DATA_ROOT/05_abruijn/20171114assembly_${name}/reads2assembly_align/bamqc/global_stats.txt
      grep "Coverage per contig" -A 1000  $DATA_ROOT/05_abruijn/20171114assembly_${name}/reads2assembly_align/bamqc/genome_results.txt > $DATA_ROOT/05_abruijn/20171114assembly_${name}/reads2assembly_align/bamqc/contig_stats.txt
 
      samtools index $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/aligned_sorted.bam
  done


##=============
##Hgap4
##=============

date=20171114
for name in cottens lapraz
   do
      mkdir -p $DATA_ROOT/04_HGAP4/${date}assembly_${name}/reads2assembly_align/bamqc
      samtools view -F 4 $DATA_ROOT/04_HGAP4/${date}assembly_${name}/reads2assembly_align/aligned_sorted.bam > $DATA_ROOT/04_HGAP4/${date}assembly_${name}/reads2assembly_align/aligned_sorted_mapped.bam
      samtools view -b -f 4 $DATA_ROOT/04_HGAP4/${date}assembly_${name}/reads2assembly_align/aligned_sorted.bam > $DATA_ROOT/04_HGAP4/${date}assembly_${name}/reads2assembly_align/aligned_sorted_unmapped.bam
      samtools view $DATA_ROOT/04_HGAP4/${date}assembly_${name}/reads2assembly_align/aligned_sorted_unmapped.bam |awk '{OFS="\t"; print ">"$1"\n"$10}' - > $DATA_ROOT/04_HGAP4/${date}assembly_${name}/reads2assembly_align/aligned_sorted_unmapped.fasta
      $QUAL_ROOT/qualimap bamqc -bam $DATA_ROOT/04_HGAP4/${date}assembly_${name}/reads2assembly_align/aligned_sorted.bam -outdir $DATA_ROOT/04_HGAP4/${date}assembly_${name}/reads2assembly_align/bamqc/ --java-mem-size=2G
      grep "Globals" -A 12  $DATA_ROOT/05_abruijn/20171114assembly_${name}/reads2assembly_align/bamqc/genome_results.txt |tail -11|sed '/^\s*$/d'|sed 's/=/\t/g'|sed -e 's/(.*%)//'|sed 's/,//g' > $DATA_ROOT/05_abruijn/20171114assembly_${name}/reads2assembly_align/bamqc/global_stats.txt
      grep "Coverage per contig" -A 1000  $DATA_ROOT/05_abruijn/20171114assembly_${name}/reads2assembly_align/bamqc/genome_results.txt > $DATA_ROOT/05_abruijn/20171114assembly_${name}/reads2assembly_align/bamqc/contig_stats.txt
   done    
   

##=============
##minimap alignmed of abruijn assembly
##=============   


for name in cottens lapraz
   do
      mkdir -p $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/bamqc
      samtools sort $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/raw_${name}_mapped_genome.sam -O BAM -o $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/raw_${name}_mapped_genome_sorted.bam
      
      samtools view -F 4 $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/raw_${name}_mapped_genome_sorted.bam > $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/raw_${name}_mapped_genome_sorted_mapped.bam
      
      samtools view -b -f 4 $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/raw_${name}_mapped_genome_sorted.bam > $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/raw_${name}_mapped_genome_sorted_unmapped.bam
      
      samtools view $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/raw_${name}_mapped_genome_sorted_unmapped.bam |awk '{OFS="\t"; print ">"$1"\n"$10}' - > $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/raw_${name}_mapped_genome_sorted_unmapped.fasta
      
      bedtools bamtofastq -i $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/raw_${name}_mapped_genome_sorted_unmapped.bam -fq $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/raw_${name}_mapped_genome_sorted_unmapped.fastq
      
      $QUAL_ROOT/qualimap bamqc -bam $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/raw_${name}_mapped_genome_sorted.bam -outdir $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/bamqc/ --java-mem-size=2G
      
      grep "Globals" -A 12  $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/bamqc/genome_results.txt |tail -11|sed '/^\s*$/d'|sed 's/=/\t/g'|sed -e 's/(.*%)//'|sed 's/,//g' > $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/bamqc/global_stats.txt
      
      grep "Coverage per contig" -A 1000  $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/bamqc/genome_results.txt > $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/bamqc/contig_stats.txt
   done 
 

```

bamqc Auswertung in R
```{r bamqc ,warning=FALSE,eval=FALSE}

bamqc_directory <- "/home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/05_abruijn/assembly_"

##------------
##global stats
##------------
for(i in c("cottens","lapraz")){
nam_0 <- paste("global_stats_abruijn_",i,sep="")

#assign(nam_0,read_delim(paste(bamqc_directory,i,"/reads2assembly_align_minimap/bamqc/global_stats.txt",sep=""), "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE))
read.delim(file=paste(bamqc_directory,i,"/reads2assembly_align_minimap/bamqc/global_stats.txt",sep=""), sep="\t")#, escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
nam <- paste("global_stats_abruijn_",i,sep="")
nam_2 <- paste("mapped_unmapped",i,sep="")

#assign(nam_2,setNames(c(as.numeric(unlist(get(nam)[3,2])),(as.numeric(unlist(get(nam)[2,2]))-as.numeric(unlist(get(nam)[3,2])))),c("mapped reads","unmapped reads")))

#assign(nam,setNames(as.vector(unlist(get(nam_2)$sequence_length_distribution[,2])),as.vector(unlist(get(nam_2)$sequence_length_distribution[,1]))))
#barplot(height =get(nam),las=2,main = nam)
}

table_stats <- data.frame("cottens"=mapped_unmappedcottens,"lapraz"=mapped_unmappedlapraz)
row.names(table_stats) <- names(mapped_unmappedcottens)
kable(table_stats)
##------------
##coverage per contig
##------------

for(i in c("cotten","lapraz")){

nam_0 <- paste("contig_stats_abruijn_",i,sep="")

assign(nam_0,read_delim(paste(bamqc_directory,i,"/reads2assembly_align_minimap/bamqc/contig_stats.txt",sep=""), "\t", escape_double = FALSE, col_names = c("blank","contig","size","bases_mapped","coverage","stdev"), trim_ws = TRUE,skip = 2))  
  
nam <- paste("contig_stats_abruijn_",i,sep="")
nam_2 <- paste("table",i,sep="")

assign(nam_2,data.frame(get(nam)[,-1]))
print(i)
kable(get(nam_2))
}
##------------
##coverage across ref
##------------

for(i in c("cottens","lapraz")){

nam_0 <- paste("coverage_stats_abruijn_",i,sep="")

assign(nam_0,read_delim(paste(bamqc_directory,i,"/reads2assembly_align_minimap/bamqc/raw_data_qualimapReport/coverage_across_reference.txt",sep=""), "\t", escape_double = FALSE,col_names = c("position","coverage"), trim_ws = TRUE,skip = 1))  
  
    
nam <- paste("coverage_stats_abruijn_",i,sep="")
nam_2 <- paste("coverage_Abruijn_",i,sep="")
nam_3 <- paste("table",i,sep="")

##cumulative genome size
assign(nam_3,cbind(get(nam_3),"location"=cumsum(get(nam_3)$size)))


assign(nam_2,data.frame(get(nam)))
ggplot(data=get(nam_2), aes(x=position, y=coverage)) +
  geom_line()+
  theme_light()+
  geom_hline(yintercept = 100,color="red",linetype="dashed")+
  geom_vline(data=get(nam_3),aes(xintercept = location),colour="blue",linetype=3)+
  annotate("text", min(get(nam_2)$position), 100, vjust = -1, label = "100 x",color="red")+
  labs(title=nam_2)
}
##------------
##mapping quality across reference
##------------

for(i in c("cottens","lapraz")){

nam_0 <- paste("mapping_stats_abruijn_",i,sep="")

assign(nam_0,read_delim(paste(bamqc_directory,i,"/reads2assembly_align_minimap/bamqc/raw_data_qualimapReport/mapping_quality_across_reference.txt",sep=""), "\t", escape_double = FALSE,col_names = c("position","MappingQuality"), trim_ws = TRUE,skip = 1))  
    
  
nam <- paste("mapping_stats_abruijn_",i,sep="")
nam_2 <- paste("mappingQuality_Abruijn_",i,sep="")
nam_3 <- paste("table",i,sep="")

assign(nam_2,data.frame(get(nam)))
ggplot(data=get(nam_2), aes(x=position, y=MappingQuality)) +
  geom_line()+
  theme_light()+
  geom_hline(yintercept = 20,color="red",linetype="dashed")+
  geom_vline(data=get(nam_3),aes(xintercept = location),colour="blue",linetype=3)+
  annotate("text", min(get(nam_2)$position), 20, vjust = -1, label = "MapQ 20",color="red")+
  labs(title=nam_2)
}

```

##Mapping all raw reads of cottens to two complete genomes

In this section I map the all raw reads from cottens back to the two completly assembled genomes of S.thermophilus and L.delbrueckii. I then continue with the reads that did not map. 

```{bash mappingreads , warning=F, eval=F}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

name=cottens
##----------------
##create database
##----------------

cat $DATA_ROOT/05_abruijn/assembly_cottens_min5000bp/fasta/multifasta/linear_1_len1924117_cov316.fa $DATA_ROOT/05_abruijn/assembly_cottens_min5000bp/fasta/multifasta/linear_2_len2265372_cov167.fa > $DATA_ROOT/05_abruijn/assembly_cottens_min5000bp/fasta/multifasta/Contig1_Contig2.fa

bwa index $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/fasta/multifasta/Contig1_Contig2.fa 

##----------------
##mapping raw reads
##----------------
mkdir -p $DATA_ROOT/05_abruijn/${date}assembly_${name}_min5000bp/reads2contigs/
bwa mem $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/fasta/multifasta/Contig1_Contig2.fa $DATA_ROOT/01_SMRTlink_data/${name}/reads.fastq -x pacbio -t 12|samtools sort -@8 -O BAM -o $DATA_ROOT/05_abruijn/${date}assembly_${name}_min5000bp/reads2contigs/aligned_sorted.bam -
 
##----------------
##converting bam file
##----------------   
      mkdir -p $DATA_ROOT/05_abruijn/${date}assembly_${name}_min5000bp/reads2contigs/bamqc
      samtools view -F 4 $DATA_ROOT/05_abruijn/${date}assembly_${name}_min5000bp/reads2contigs/aligned_sorted.bam > $DATA_ROOT/05_abruijn/${date}assembly_${name}_min5000bp/reads2contigs/aligned_sorted_mapped.bam 
      samtools view -b -f 4 $DATA_ROOT/05_abruijn/${date}assembly_${name}_min5000bp/reads2contigs/aligned_sorted.bam > $DATA_ROOT/05_abruijn/${date}assembly_${name}_min5000bp/reads2contigs/aligned_sorted_unmapped.bam
      bedtools bamtofastq -i $DATA_ROOT/05_abruijn/${date}assembly_${name}_min5000bp/reads2contigs/aligned_sorted_unmapped.bam -fq $DATA_ROOT/05_abruijn/${date}assembly_${name}_min5000bp/reads2contigs/aligned_sorted_unmapped.fastq
      $QUAL_ROOT/qualimap bamqc -bam $DATA_ROOT/05_abruijn/${date}assembly_${name}_min5000bp/reads2contigs/aligned_sorted.bam -outdir $DATA_ROOT/05_abruijn/${date}assembly_${name}_min5000bp/reads2contigs/bamqc/ --java-mem-size=2G
      grep "Globals" -A 12  $DATA_ROOT/05_abruijn/20171114assembly_${name}_min5000bp/reads2contigs/bamqc/genome_results.txt |tail -11|sed '/^\s*$/d'|sed 's/=/\t/g'|sed -e 's/(.*%)//'|sed 's/,//g' > $DATA_ROOT/05_abruijn/${date}assembly_${name}_min5000bp/reads2contigs/bamqc/global_stats.txt
      grep "Coverage per contig" -A 1000  $DATA_ROOT/05_abruijn/${date}assembly_${name}_min5000bp/reads2assembly_align/bamqc/genome_results.txt > $DATA_ROOT/05_abruijn/${date}assembly_${name}_min5000bp/reads2assembly_align/bamqc/contig_stats.txt
 


```
##Blasting all unmapped reads to the entire blast db

To insure that we are able to assemble the large majority of the diversity we want to check in a following the residual reads which have not been able to map back to the assembled contigs. 


```{bash blast_unmapped , warning=FALSE, eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh


##----------------
##ncbi download
##----------------

##curl --proxy proxy-bvcol.admin.ch:8080 -O ftp://ftp.ncbi.nlm.nih.gov/blast/db/FASTA/nt.gz
##gunzip -c $aditiBlastDB/nt.gz > aditiBlastDB/nt_db/
##makeblastdb -in $aditiBlastDB/nt -parse_seqids -dbtype nucl
##---------------- 
##blast
##----------------
name=lapraz
for name in cottens lapraz
   do
    mkdir -p $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/blast_unmapped/

    $blast_root/blastn -num_threads 15 -max_hsps 1 -max_target_seqs 1 -task megablast -show_gis -query $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/reads2assembly_align/aligned_sorted_unmapped.fasta -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle slen" -db $aditiBlastDB/nt -out $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/blast_unmapped/blastedUnmapped.txt -evalue 0.01 -word_size 64
  done
  
  
##---------------- 
##blast
##----------------
name=cottens
for name in cottens lapraz
   do
    mkdir -p $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/blast_unmapped/

    $blast_root/blastn -num_threads 15 -max_hsps 1 -max_target_seqs 1 -task megablast -show_gis -query     $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/reads2assembly_align/aligned_sorted_unmapped.fasta -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle slen" -db $aditiBlastDB/nt -out $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/blast_unmapped/blastedUnmapped_wordsize16.txt -evalue 0.01 -word_size 16
  done  

```

##looking at the blast data
In the following we look specifically at the reads which did not map back assembled contigs. We blasted these reads against the entire NCBI database. 

```{bash blast_unmapped_output_bash ,eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

##----------------
##blast output
##----------------
for name in cottens lapraz
   do
    echo 'The following number of reads did not map back to the ' ${name} 'assembly :'
    grep ">" -c $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/reads2assembly_align/aligned_sorted_unmapped.fasta |cut -f 1
    echo 'From these the following number of reads had a blast hit'
    wc -l $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/blast_unmapped/blastedUnmapped.txt |cut -f1

    cut -f 13 $DATA_ROOT/05_abruijn/assembly_cottens_min5000bp/blast_unmapped/blastedUnmapped.txt |sort|uniq -c
   done
```
 
##Reads which neither map to assembled contigs nor blast

We identified in the previous step a few reads which did not map to the assembled contigs nor showed a blast hit against the entire ncbi database.
In the following we try to extract these reads and do the following.
1. check GC- content of reads
2. look a sequencing score of the reads 

```{bash unmapped_unblasted_reads,eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

#----------------
##reads that had blast hits
##----------------
#name=cottens
for name in cottens lapraz
   do
    grep ">" $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/reads2assembly_align/aligned_sorted_unmapped.fasta|sed 's/>//g' > $DATA_ROOT/05_abruijn/assembly_cottens_min5000bp/reads2assembly_align/aligned_sorted_unmapped_names.txt
    cat $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/blast_unmapped/blastedUnmapped.txt |cut -f 1 > $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/blast_unmapped/blastedUnmapped_with_blastHits.txt
  done
  

#----------------
##diff
##----------------
for name in cottens lapraz
   do
    diff $DATA_ROOT/05_abruijn/assembly_cottens_min5000bp/reads2assembly_align/aligned_sorted_unmapped_names.txt $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/blast_unmapped/blastedUnmapped_with_blastHits.txt |grep "<"|sed 's/< //g' > $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/blast_unmapped/blastedUnmapped_without_blastHits.txt
  done 
  
#----------------
##creat fasta of diff for GC content
##----------------
for name in cottens lapraz
   do
    rm $DATA_ROOT/05_abruijn/assembly_cottens_min5000bp/reads2assembly_align/aligned_sorted_unmapped_unblasted.fasta
    for query in $(cat $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/blast_unmapped/blastedUnmapped_without_blastHits.txt)
      do
        grep ${query} -A 1 $DATA_ROOT/05_abruijn/assembly_cottens_min5000bp/reads2assembly_align/aligned_sorted_unmapped.fasta >> $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/reads2assembly_align/aligned_sorted_unmapped_unblasted.fasta
        #echo ${query}
      done
    infoseq -auto $DATA_ROOT/05_abruijn/assembly_cottens_min5000bp/reads2assembly_align/aligned_sorted_unmapped_unblasted.fasta > $DATA_ROOT/05_abruijn/assembly_${name}_min5000bp/reads2assembly_align/aligned_sorted_unmapped_unblasted_seqinfo.txt
   done

#----------------
##calculate gc content (infoseq)
##----------------
for name in cottens lapraz
   do
    infoseq -nocolumns -delimiter , -only -name -length -pgc $DATA_ROOT/01_SMRTlink_data/${name}/reads.fastq > $DATA_ROOT/01_SMRTlink_data/${name}/reads_infoseq_gc_length.txt
   done
    
#----------------
##number of subreads
##----------------

##movie name    
m54073_171026_034928 ##lapaz
m54073_171026_135911 ##cottens
##bam file location
/home/bioinf/bioinf_archive/32_scmi_storage/archivePacBio/fettsirte1_02112017/scratch/pacbio_tmp/tmpWL3QGn/1_8_La_praz_3_D02.tar.gz/m54073_171026_034928.subreads.bam
/home/bioinf/bioinf_archive/32_scmi_storage/archivePacBio/fettsirte1_02112017/scratch/pacbio_tmp/tmpWL3QGn/2_7_Cottens_4_E02.tar.gz/m54073_171026_135911.subreads.bam

##lapraz
samtools view /home/bioinf/bioinf_archive/32_scmi_storage/archivePacBio/fettsirte1_02112017/scratch/pacbio_tmp/tmpWL3QGn/1_8_La_praz_3_D02.tar.gz/m54073_171026_034928.subreads.bam|grep "m54073_171026_034928" -c
#763335

##cottens
samtools view /home/bioinf/bioinf_archive/32_scmi_storage/archivePacBio/fettsirte1_02112017/scratch/pacbio_tmp/tmpWL3QGn/2_7_Cottens_4_E02.tar.gz/m54073_171026_135911.subreads.bam|grep "m54073_171026_135911" -c
##379465

```

```{r unmapped_unblasted_reads ,warning=FALSE,eval=FALSE}
##--------------
##import files cottens 
##--------------

blastedUnmapped_with_blastHits_cottens <- read_csv("~/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/05_abruijn/assembly_cottens_min5000bp/blast_unmapped/blastedUnmapped_with_blastHits.txt",  col_names = "unmapped_withBlastHits")

blastedUnmapped_without_blastHits_cottens <- read_csv("~/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/05_abruijn/assembly_cottens_min5000bp/blast_unmapped/blastedUnmapped_without_blastHits.txt",  col_names = "unmapped_withoutBlastHits")

infoseq_all_cottens <- read_csv("~/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/01_SMRTlink_data/cottens/reads_infoseq_gc_length.txt") 
View(infoseq_all_cottens)

##--------------
##import files lapraz 
##--------------

blastedUnmapped_with_blastHits_lapraz <- read_csv("~/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/05_abruijn/assembly_lapraz_min5000bp/blast_unmapped/blastedUnmapped_with_blastHits.txt",  col_names = "unmapped_withBlastHits")

blastedUnmapped_without_blastHits_lapraz <- read_csv("~/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/05_abruijn/assembly_lapraz_min5000bp/blast_unmapped/blastedUnmapped_without_blastHits.txt",  col_names = "unmapped_withoutBlastHits")

infoseq_all_lapraz <- read_csv("~/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/01_SMRTlink_data/lapraz/reads_infoseq_gc_length.txt") 
View(infoseq_all_lapraz)


```



##ReAssembly of all non-Genome reads

In the following we reassemble all reads that did not map back to the completetly assembled genomes of S.thermophilus and L.del. 

```{bash asembly_plasmids , warning=F, eval=F}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh
cov=200
##----------------
##abruijn assembly plasmids
##----------------
for name in cottens lapraz
   do
    mkdir -p $DATA_ROOT/13_abruijn_plasmidAssembly/${name}_unmappedminimap_assembly/{fasta,blast,quast,coverage}
   $abruijn_root/abruijn -t 10 $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/raw_${name}_mapped_genome_sorted_unmapped.fastq $DATA_ROOT/13_abruijn_plasmidAssembly/${name}_unmappedminimap_assembly/fasta ${cov} -p pacbio -i 2 -o 2500



##----------------
##quast
##----------------
    quast.py $DATA_ROOT/13_abruijn_plasmidAssembly/${name}_unmappedminimap_assembly/fasta/polished_1.fasta -t 8 -o $DATA_ROOT/13_abruijn_plasmidAssembly/${name}_unmappedminimap_assembly/quast/
##---------------- 
##blast
##----------------
$blast_root/blastn -num_threads 16 -max_hsps 1 -max_target_seqs 1 -task megablast -show_gis -query $DATA_ROOT/13_abruijn_plasmidAssembly/${name}_unmappedminimap_assembly/fasta/polished_1.fasta -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle slen" -db $blastdb/all_prokaryotes -out $DATA_ROOT/13_abruijn_plasmidAssembly/${name}_unmappedminimap_assembly/blastAssembly.txt -evalue 0.01 -word_size 64

##----------------
##coverage
##----------------

    $bbmap_root/bbwrap.sh ref=$DATA_ROOT/13_abruijn_plasmidAssembly/${name}_unmappedminimap_assembly/fasta/polished_1.fasta 
        $bbmap_root/bbwrap.sh ref=$DATA_ROOT/13_abruijn_plasmidAssembly/${name}_unmappedminimap_assembly/fasta/polished_1.fasta nodisk in=10 $DATA_ROOT/05_abruijn/assembly_${name}/reads2assembly_align_minimap/raw_${name}_mapped_genome_sorted_unmapped.fastq out=$DATA_ROOT/13_abruijn_plasmidAssembly/${name}_unmappedminimap_assembly/coverage 

done
```


##functional pathway annotation with HuMaNn2

HUMAnN is a pipeline for efficiently and accurately profiling the presence/absence and abundance of microbial pathways in a community from metagenomic or metatranscriptomic sequencing data (typically millions of short DNA/RNA reads).

```{bash humann2 ,eval=FALSE}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh
threads=16

##=============
##make database
##=============
mkdir -p $DATA_ROOT/12_humann2/db
#humann2_databases --download uniref uniref90_diamond $DATA_ROOT/12_humann2/db
#humann2_databases --download chocophlan full $DATA_ROOT/12_humann2/db

#humann2_config --update database_folders protein $DATA_ROOT/12_humann2/db/uniref/

##=============
##run humann2 on raw reads
##=============
date=20171114
for name in cottens lapraz
   do
      mkdir -p $DATA_ROOT/12_humann2/{db,$name_rawReads}
     humann2 --threads ${threads} --nucleotide-database $DATA_ROOT/12_humann2/db/chocophlan/ --protein-database $DATA_ROOT/12_humann2/db/uniref/ --metaphlan $metapha/metaphlan2.py --taxonomic-profile $DATA_ROOT/06_metaphlan/${name}_raw/metaphlan_profile/profiled_metagenome_${name}_raw.txt --diamond /home/bioinf/extradrv --input $DATA_ROOT/01_SMRTlink_data/${name}/preassembled_reads/preads4falcon_${name}.fasta --output $OUTPUT_DIR $DATA_ROOT/12_humann2/${name}_rawReads/
   done  

```
Every humann2 mapping step (1. Chocophlan,Metaphlan,proteindb,..) is done with bowtie and parameter settings for Illumina reads. 

##Blasting raw reads against prokaryotic db

A second possibility is to BLAST all raw reads against the entire prokaryotic blastdb.
By only ouputing the best hit we can get a good feel for the diversity of microorganisms therein. 
```{bash blast_raw_reads , warning=F, eval=F}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

###-----------------
##filter for long reads (>10'000 bp)
###-----------------
date=20171114
for name in cottens lapraz
   do
    prinseq-lite.pl -fastq $DATA_ROOT/01_SMRTlink_data/${name}/reads.fastq -min_len 10000 -out_good $DATA_ROOT/01_SMRTlink_data/${name}/reads_min10000bp -out_bad null -out_format 1 
   done
###-----------------
##blast raw reads
###-----------------

date=20171114
for name in cottens lapraz
   do
     mkdir -p $DATA_ROOT/01_SMRTlink_data/${name}/blast_min10000bp/
     $blast_root/blastn -num_threads 6 -max_hsps 1 -max_target_seqs 1 -task megablast -show_gis -query $DATA_ROOT/01_SMRTlink_data/${name}/reads_min10000bp.fasta -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle slen" -db $blastdb/all_prokaryotes -out $DATA_ROOT/01_SMRTlink_data/${name}/blast_min10000bp/blast_rawReads.txt -evalue 0.01 -word_size 64
   done
   
###-----------------
##see mapping regions
###-----------------
   
date=20171114
for name in cottens lapraz
   do
      echo species level for ${name}
           sort -k1,1 -k12,12nr -k11,11n  $DATA_ROOT/01_SMRTlink_data/${name}/blast_min10000bp/blast_rawReads.txt | sort -u -k1,1 --merge|cut -f 13|awk 'BEGIN {OFS=" "}; {print $1, $2}'|sort|uniq -c
      echo strain level for ${name}
      sort -k1,1 -k12,12nr -k11,11n  $DATA_ROOT/01_SMRTlink_data/${name}/blast_min10000bp/blast_rawReads.txt | sort -u -k1,1 --merge|cut -f 13|sort| uniq -c
   done


```



##compute alignment depth along ref genomes

In order to control the completness of the metagenome assembly we create a nucleotide wise coverage along all matching reference genomes.
We do this with a samtools depth command. 

```{bash alignment_depth , warning=F, eval=F}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

###-----------------
##Hgap4 depth
###-----------------
date=20171114
for name in cottens lapraz
   do
    samtools depth -a -Q 10 $DATA_ROOT/04_HGAP4/${date}assembly_${name}/reads2assembly_align/aligned_sorted.bam > $DATA_ROOT/04_HGAP4/${date}assembly_${name}/reads2assembly_align/aligned_sorted_coverage.bed
  done
  
###-----------------
##Abruijn depth
###-----------------
date=20171114
for name in cottens lapraz
   do
    samtools depth -a -Q 10 $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/aligned_sorted.bam > $DATA_ROOT/05_abruijn/${date}assembly_${name}/reads2assembly_align/aligned_sorted_coverage.bed
  done  
```



##Metaquast
We can analyse the current assemblages with a tool specifically designed for metagenomics.
If no reference genomes are provided it will BLASTn against the SILVA database to identify potential reference candidates. 
```{bash metaquast , warning=F, eval=F}
source /home/bioinf/bioinf_archive/43_sovi_storage/Projects/Fettsirte/Run20171109/02_scripts/fileShortcuts.sh

##=============
##Abruijn assembly
##=============

date=20171114
for name in cottens lapraz
   do
     mkdir -p $DATA_ROOT/05_abruijn/${date}assembly_${name}/metaquast/
     sudo metaquast.py -t 2 $DATA_ROOT/05_abruijn/assembly_${name}/fasta/polished_1.fasta --max-ref-number 40 -o $DATA_ROOT/05_abruijn/${date}assembly_${name}/metaquast/
   done

##=============
##Hgap4 assembly
##=============

date=20171114
for name in cottens lapraz
   do
     mkdir -p $DATA_ROOT/04_HGAP4/${date}assembly_${name}/metaquast/
     sudo metaquast.py -t 12 $DATA_ROOT/04_HGAP4/${date}assembly_${name}/fasta/assembly.fasta --max-ref-number 40 -o $DATA_ROOT/04_HGAP4/${date}assembly_${name}/metaquast/
   done
   
##=============
##Pre-Assembled reads
##=============
     sudo metaquast.py -t 12 01_SMRTlink_data/${name}/preassembled_reads/preads4falcon.fasta --max-ref-number 40 -o $DATA_ROOT/07_PreAssembled/${date}assembly_${name}/metaquast/

     
   

date=20171114
for name in cottens lapraz
   do
     mkdir -p $DATA_ROOT/07_PreAssembled/${date}assembly_${name}/metaquast/
     #sudo metaquast.py -t 12 DATA_ROOT/01_SMRTlink_data/${name}/preassembled_reads/preads4falcon.fasta --max-ref-number 40 -o $DATA_ROOT/07_PreAssembled/${date}assembly_${name}/metaquast/
     #sudo metaquast.py -t 12 DATA_ROOT/01_SMRTlink_data/${name}/preassembled_reads/preads4falcon.fasta --references-list $DATA_ROOT/01_SMRTlink_data/reference_list.txt -o $DATA_ROOT/07_PreAssembled/${date}assembly_${name}/metaquast/
          sudo metaquast.py -t 12 $DATA_ROOT/01_SMRTlink_data/${name}/preassembled_reads/preads4falcon.fasta -R $DATA_ROOT/08_metaquast/01_referenceData/lactobabillus_helveticus.fasta,$DATA_ROOT/08_metaquast/01_referenceData/lactobaicllus_delbruckii.fasta,$DATA_ROOT/08_metaquast/01_referenceData/lactobaicllus_delbruckii.fasta -o $DATA_ROOT/07_PreAssembled/${date}assembly_${name}/metaquast/
   done     

```



